// @node-projects/web-component-designer/dist/elements/helper/ElementHelper.js
export function newElementFromString(text) {
    const range = document.createRange();
    range.selectNode(document.body);
    const fragment = range.createContextualFragment(text);
    return fragment.firstChild;
}
export var ElementDisplayType;
(function (ElementDisplayType) {
    ElementDisplayType[ElementDisplayType["none"] = 0] = "none";
    ElementDisplayType[ElementDisplayType["inline"] = 1] = "inline";
    ElementDisplayType[ElementDisplayType["block"] = 2] = "block";
})(ElementDisplayType || (ElementDisplayType = {}));
export function isInline(element) {
    return element != null && window.getComputedStyle(element).display.startsWith('inline');
}
export function getElementDisplaytype(element) {
    const display = window.getComputedStyle(element).display;
    return display == 'none' ? ElementDisplayType.none : display.startsWith('inline') ? ElementDisplayType.inline : ElementDisplayType.block;
}
export function isEmptyTextNode(node) {
    return node.textContent.trim() == '' && node.textContent.indexOf('\xa0' /* &nbsp; */) < 0;
}
