// @node-projects/web-component-designer/dist/elements/helper/Helper.js
export function isAppleDevice() {
    return window.navigator.platform?.startsWith("Mac") || window.navigator.platform === "iPhone" || window.navigator.platform === "iPad" || window.navigator.platform === "iPod";
}
export function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}
export async function exportData(blob, fileName) {
    const url = URL.createObjectURL(blob);
    const a = document.createElement('a');
    a.href = url;
    a.style.display = 'none';
    a.download = fileName;
    document.body.appendChild(a);
    a.click();
    document.body.removeChild(a);
    await sleep(300);
}
export function dataURItoBlob(dataURI) {
    var mime = dataURI.split(',')[0].split(':')[1].split(';')[0];
    var binary = atob(dataURI.split(',')[1]);
    var array = [];
    for (var i = 0; i < binary.length; i++) {
        array.push(binary.charCodeAt(i));
    }
    return new Blob([new Uint8Array(array)], { type: mime });
}
export function pointInRect(point, rect) {
    return point.x >= rect.x && point.x <= rect.x + rect.width && point.y >= rect.y && point.y <= rect.y + rect.height;
}
export function removeTrailing(text, char) {
    if (text.endsWith(char ?? '/'))
        return text.substring(0, text.length - 1);
    return text;
}
export function removeLeading(text, char) {
    if (text.startsWith(char ?? '/'))
        return text.substring(1);
    return text;
}
export function requestAnimationFramePromise() {
    return new Promise(resolve => requestAnimationFrame(resolve));
}
export function arraysEqual(a, b) {
    if (a === b)
        return true;
    if (a == null || b == null)
        return false;
    if (a.length !== b.length)
        return false;
    for (var i = 0; i < a.length; ++i) {
        if (a[i] !== b[i])
            return false;
    }
    return true;
}
