// @node-projects/web-component-designer/dist/elements/helper/SelectionHelper.js
export function shadowrootGetSelection(shadowRoot) {
    let selection = document.getSelection();
    if (selection.getComposedRanges)
        selection = selection.getComposedRanges(shadowRoot);
    else if (shadowRoot.getSelection)
        selection = shadowRoot.getSelection();
    return selection;
}
function wrapTextNodesInSpan(range, spans) {
    function wrapNode(node) {
        const span = document.createElement('span');
        spans.push(span);
        node.parentNode.insertBefore(span, node);
        span.appendChild(node);
    }
    function processNode(node) {
        switch (node.nodeType) {
            case Node.TEXT_NODE:
                wrapNode(node);
                break;
            case Node.DOCUMENT_FRAGMENT_NODE:
            case Node.ELEMENT_NODE:
                Array.from(node.childNodes).forEach(processNode);
                break;
        }
    }
    const fragment = range.extractContents();
    processNode(fragment);
    range.insertNode(fragment);
}
function staticRangeToRange(staticRange) {
    const range = document.createRange();
    range.setStart(staticRange.startContainer, staticRange.startOffset);
    range.setEnd(staticRange.endContainer, staticRange.endOffset);
    return range;
}
export function wrapSelectionInSpans(selection) {
    const spans = [];
    if (selection[0] instanceof StaticRange) {
        wrapTextNodesInSpan(staticRangeToRange(selection[0]), spans);
    }
    else {
        if (!selection.rangeCount)
            return spans;
        const range = selection.getRangeAt(0);
        wrapTextNodesInSpan(range, spans);
    }
    if (selection.removeAllRanges)
        selection.removeAllRanges();
    return spans;
}
