// @node-projects/web-component-designer/dist/elements/widgets/designerView/extensions/contextMenu/CollaborationCommentsContextMenu.js
import { getCollaborationNodeIndex } from '../../../../services/collaborationService/CollaborationNodeIndex.js';
function truncateText(text, maxLength) {
    if (!text || text.length <= maxLength)
        return text;
    return text.substring(0, Math.max(0, maxLength - 1)) + '…';
}
function createCommentId() {
    return globalThis.crypto?.randomUUID?.() ?? `comment-${Date.now()}-${Math.floor(Math.random() * 100000)}`;
}
export class CollaborationCommentsContextMenu {
    shouldProvideContextmenu(event, designerView, designItem, initiator) {
        return !!designerView.instanceServiceContainer.collaborationService && !!designItem && !designItem.isRootItem;
    }
    provideContextMenuItems(event, designerCanvas, designItem) {
        const collaborationService = designerCanvas.instanceServiceContainer.collaborationService;
        const targetNodeIndex = getCollaborationNodeIndex(designItem);
        if (targetNodeIndex == null)
            return [];
        const comments = collaborationService.comments.filter(x => x.targetNodeIndex === targetNodeIndex);
        const localPeerId = collaborationService.session?.peerId ?? 'local';
        const items = [
            {
                title: 'add comment',
                action: () => {
                    const value = globalThis.prompt?.('Add comment', '')?.trim();
                    if (!value)
                        return;
                    const now = Date.now();
                    collaborationService.upsertComment({
                        id: createCommentId(),
                        text: value,
                        authorPeerId: localPeerId,
                        targetNodeIndex,
                        targetDesignItemId: designItem.id,
                        createdAt: now,
                        updatedAt: now,
                        resolved: false
                    });
                }
            }
        ];
        if (comments.length) {
            items.push({ title: '-' });
            items.push({
                title: `comments (${comments.length})`,
                children: comments.map(comment => this.createCommentMenuItem(comment, collaborationService, designItem))
            });
        }
        return items;
    }
    createCommentMenuItem(comment, collaborationService, designItem) {
        return {
            title: `${comment.resolved ? 'resolved' : 'open'}: ${truncateText(comment.text, 36)}`,
            children: [
                {
                    title: comment.resolved ? 'reopen' : 'resolve',
                    action: () => collaborationService.upsertComment({
                        ...comment,
                        targetDesignItemId: designItem.id,
                        resolved: !comment.resolved,
                        updatedAt: Date.now()
                    })
                },
                {
                    title: 'remove',
                    action: () => collaborationService.removeComment(comment.id)
                }
            ]
        };
    }
}
