// @node-projects/web-component-designer/dist/elements/widgets/designerView/extensions/contextMenu/MultipleItemsSelectedContextMenu.js
import { switchContainer } from '../../../../helper/SwitchContainerHelper.js';
import { DesignItem } from '../../../../item/DesignItem.js';
export class MultipleItemsSelectedContextMenu {
    orderIndex = 60;
    shouldProvideContextmenu(event, designerCanvas, designItem, initiator) {
        if (designItem.instanceServiceContainer.selectionService.selectedElements.length > 1) {
            return true;
        }
        return false;
    }
    provideContextMenuItems(event, designerCanvas, designItem) {
        return [
            {
                title: 'wrap in',
                children: [
                    {
                        title: 'div',
                        action: () => {
                            const grp = designItem.openGroup("wrap in Div");
                            let elements = designItem.instanceServiceContainer.selectionService.selectedElements;
                            let newContainer = document.createElement('div');
                            const newContainerDesignItem = DesignItem.createDesignItemFromInstance(newContainer, designItem.serviceContainer, designItem.instanceServiceContainer);
                            elements[0].insertAdjacentElement(newContainerDesignItem, 'beforebegin');
                            switchContainer(elements, newContainerDesignItem, true, 10);
                            grp.commit();
                            designItem.instanceServiceContainer.selectionService.setSelectedElements([newContainerDesignItem]);
                        }
                    }
                ]
            }
        ];
    }
}
