// @node-projects/web-component-designer/dist/elements/widgets/designerView/extensions/contextMenu/PasteFormatContextMenu.js
import { NodeType } from '../../../../item/NodeType.js';
import { createPasteFormatSnapshotFromEntries, getPasteFormatEntries } from '../../../../services/copyPasteService/PasteFormatSnapshot.js';
export class PasteFormatContextMenu {
    shouldProvideContextmenu(event, designerCanvas, designItem, initiator) {
        return !designerCanvas.readOnly
            && this._getTargetDesignItems(designerCanvas).length > 0;
    }
    provideContextMenuItems(event, designerCanvas, designItem) {
        return [{
                title: 'paste format',
                children: [
                    this._createPasteFormatItem(designerCanvas, 'all'),
                    { title: '-' },
                    this._createPasteFormatItem(designerCanvas, 'border'),
                    this._createPasteFormatItem(designerCanvas, 'background'),
                    this._createPasteFormatItem(designerCanvas, 'text')
                ]
            }];
    }
    _createPasteFormatItem(designerCanvas, kind) {
        return {
            title: kind,
            action: () => {
                void this._pasteFormat(designerCanvas, kind);
            }
        };
    }
    _getTargetDesignItems(designerCanvas) {
        return designerCanvas.instanceServiceContainer.selectionService.selectedElements?.filter(x => x?.nodeType === NodeType.Element && !x.isRootItem) ?? [];
    }
    async _pasteFormat(designerCanvas, kind) {
        const [pasteDesignItems] = await designerCanvas.serviceContainer.copyPasteService.getPasteItems(designerCanvas.serviceContainer, designerCanvas.instanceServiceContainer);
        const pasteSource = pasteDesignItems?.find(x => x?.nodeType === NodeType.Element && x.hasStyles && !x.isRootItem);
        const snapshot = pasteSource ? createPasteFormatSnapshotFromEntries(pasteSource.styles()) : null;
        if (!snapshot) {
            return;
        }
        const entries = getPasteFormatEntries(snapshot, kind);
        const targets = this._getTargetDesignItems(designerCanvas);
        if (!entries.length || !targets.length) {
            return;
        }
        const group = targets[0].openGroup(`paste format: ${kind}`);
        let changed = false;
        for (const target of targets) {
            for (const entry of entries) {
                if (target.getStyle(entry.name) === entry.value) {
                    continue;
                }
                target.setStyle(entry.name, entry.value);
                changed = true;
            }
        }
        if (changed) {
            group.commit();
        }
        else {
            group.abort();
        }
    }
}
