// @node-projects/web-component-designer/dist/elements/services/propertiesService/services/CommonPropertiesService.js
import { PropertyType } from '../PropertyType.js';
import { AbstractPropertiesService } from './AbstractPropertiesService.js';
export class CommonPropertiesService extends AbstractPropertiesService {
    listNeedsRefresh(designItem) {
        return false;
    }
    //@ts-ignore
    commonProperties = [
        {
            name: "id",
            type: "string",
            service: this,
            propertyType: PropertyType.propertyAndAttribute
        }, {
            name: "class",
            type: "string",
            service: this,
            attributeName: "class",
            propertyName: "className",
            propertyType: PropertyType.propertyAndAttribute
        }, {
            name: "title",
            type: "string",
            service: this,
            propertyType: PropertyType.propertyAndAttribute
        }, {
            name: "tabindex",
            type: "number",
            service: this,
            propertyType: PropertyType.propertyAndAttribute
        }
    ];
    name = "common";
    isHandledElement(designItem) {
        return true;
    }
    getProperty(designItem, name) {
        return this.commonProperties[name];
    }
    getProperties(designItem) {
        return this.commonProperties;
    }
}
