// @node-projects/web-component-designer/dist/elements/services/propertiesService/services/ListPropertiesService.js
import { AbstractPropertiesService } from './AbstractPropertiesService.js';
import { PropertyType } from '../PropertyType.js';
import { RefreshMode } from '../IPropertiesService.js';
export class ListPropertiesService extends AbstractPropertiesService {
    getRefreshMode(designItem) {
        return RefreshMode.full;
    }
    name = "list";
    _propertys = new Map();
    constructor(propertyDefinitions) {
        super();
        for (let e in propertyDefinitions) {
            let parr = [];
            this._propertys.set(e, parr);
            for (let p in propertyDefinitions[e]) {
                let pdef = propertyDefinitions[e][p];
                parr.push({
                    name: pdef.name,
                    propertyName: pdef.propertyName,
                    attributeName: pdef.attributeName,
                    description: pdef.description,
                    type: pdef.type,
                    default: pdef.default,
                    min: pdef.min,
                    max: pdef.max,
                    step: pdef.step,
                    values: pdef.values,
                    enumValues: pdef.enumValues,
                    value: pdef.value,
                    defaultValue: pdef.defaultValue,
                    service: this,
                    propertyType: pdef.propertyType ?? PropertyType.propertyAndAttribute
                });
            }
        }
    }
    isHandledElement(designItem) {
        return this._propertys.has(designItem.element.localName);
    }
    _notifyChangedProperty(designItem, property, value) {
    }
    getProperties(designItem) {
        return this._propertys.get(designItem.element.localName);
    }
}
