// @node-projects/css-parser/dist/type.js
export var CssTypes;
(function (CssTypes) {
    CssTypes["stylesheet"] = "stylesheet";
    CssTypes["rule"] = "rule";
    CssTypes["declaration"] = "declaration";
    CssTypes["comment"] = "comment";
    CssTypes["whitespace"] = "whitespace";
    CssTypes["atRule"] = "at-rule";
    CssTypes["container"] = "container";
    CssTypes["charset"] = "charset";
    CssTypes["counterStyle"] = "counter-style";
    CssTypes["document"] = "document";
    CssTypes["customMedia"] = "custom-media";
    CssTypes["fontFace"] = "font-face";
    CssTypes["fontFeatureValues"] = "font-feature-values";
    CssTypes["host"] = "host";
    CssTypes["import"] = "import";
    CssTypes["keyframes"] = "keyframes";
    CssTypes["keyframe"] = "keyframe";
    CssTypes["layer"] = "layer";
    CssTypes["media"] = "media";
    CssTypes["namespace"] = "namespace";
    CssTypes["page"] = "page";
    CssTypes["pageMarginBox"] = "page-margin-box";
    CssTypes["positionTry"] = "position-try";
    CssTypes["property"] = "property";
    CssTypes["scope"] = "scope";
    CssTypes["startingStyle"] = "starting-style";
    CssTypes["supports"] = "supports";
    CssTypes["viewTransition"] = "view-transition";
})(CssTypes || (CssTypes = {}));
