// @node-projects/web-component-designer/dist/elements/widgets/bindings/BindingsEditor.js
import { css, BaseCustomWebComponentConstructorAppend, html } from '@node-projects/base-custom-webcomponent';
export var targetType;
(function (targetType) {
    targetType["property"] = "property";
    targetType["style"] = "style";
    targetType["attribute"] = "attribute";
    targetType["event"] = "event";
})(targetType || (targetType = {}));
export var valueType;
(function (valueType) {
    valueType["string"] = "string";
    valueType["number"] = "number";
    valueType["boolean"] = "boolean";
})(valueType || (valueType = {}));
export class BindingsEditor extends BaseCustomWebComponentConstructorAppend {
    static style = css `
    * { font-size: 16px; }
    .list { 
      display: flex;
      flex-direction: column;
      gap: 5px;
    }
    `;
    static template = html `
   <div class="list" style="">
    <span>Expression</span>
    <div>
      <input style="width:100%;">
      <button>...</button>
    </div>
    <!--<span>Target</span>
    <select id="target">
        <option>Property</option>
        <option>Attribute</option>
        <option>Style</option>
        <option>Event</option>
    </select>-->
    <div>
      <input id="mode" type="checkbox"><label for="mode">Two Way</label>    
    </div>
    <div>
      <input id="inverted" type="checkbox"><label for="inverted">Inverted</label>   
    </div>
    <div>
      <input id="nullSafe" type="checkbox"><label for="nullSafe">Null Safe</label>   
    </div>
</div>
  `;
    constructor(targetType) {
        super();
    }
}
customElements.define('node-projects-bindings-editor', BindingsEditor);
