// @node-projects/base-custom-webcomponent/dist/DomHelper.js
export class DomHelper {
    static *getAllChildNodes(element, includeShadowDom = false, ignore) {
        if (!ignore || ignore.indexOf(element) < 0) {
            if (element.children) {
                for (const node of element.children) {
                    yield node;
                    if (includeShadowDom && node.shadowRoot != null) {
                        yield node.shadowRoot;
                        const childs = DomHelper.getAllChildNodes(node.shadowRoot, includeShadowDom, ignore);
                        for (const cnode of childs) {
                            yield cnode;
                        }
                    }
                    const childs = DomHelper.getAllChildNodes(node, includeShadowDom, ignore);
                    for (const cnode of childs) {
                        yield cnode;
                    }
                }
            }
            if (includeShadowDom && element.shadowRoot != null) {
                yield element.shadowRoot;
                const childs = DomHelper.getAllChildNodes(element.shadowRoot, includeShadowDom);
                for (const cnode of childs) {
                    yield cnode;
                }
            }
        }
        return null;
    }
    static removeAllChildnodes(node, className) {
        if (!className) {
            for (let c = node.firstChild; c !== null; c = node.firstChild) {
                node.removeChild(c);
            }
        }
        else {
            const elements = node.querySelectorAll('.' + className);
            for (const e of elements) {
                if (e.parentNode == node)
                    node.removeChild(e);
            }
        }
    }
    static nodeIndex(node) {
        let i = 0;
        while ((node = node.previousSibling) != null)
            i++;
        return i;
    }
    static childIndex(node) {
        for (let i = 0; i < node.parentElement.children.length; i++)
            if (node.parentElement.children[i] == node)
                return i;
        return -1;
    }
    static getHost(node) {
        while (node.parentElement)
            node = node.parentElement;
        if (node.host)
            return node.host;
        return node.parentNode.host;
    }
    static nodeIsChildOf(node, parentNode) {
        while (node.parentElement) {
            if (node.parentElement == parentNode)
                return true;
            node = node.parentElement;
        }
        return false;
    }
}
