// @node-projects/web-component-designer/dist/elements/services/propertiesService/propertyEditors/BasePropertyEditor.js
import { ValueType } from '../ValueType.js';
export class BasePropertyEditor {
    element;
    property;
    designItems;
    disableChangeNotification = false;
    constructor(property) {
        this.property = property;
    }
    async _valueChanged(newValue) {
        if (!this.disableChangeNotification) {
            if (this.designItems && this.designItems.length) {
                const cg = this.designItems[0].openGroup("set property: " + this.property.name);
                if (newValue == null)
                    this.property.service.clearValue(this.designItems, this.property, 'value');
                else
                    await this.property.service.setValue(this.designItems, this.property, newValue);
                cg.commit();
            }
        }
    }
    designItemsChanged(designItems) {
        this.designItems = designItems;
    }
    refreshValueWithoutNotification(valueType, value) {
        if (valueType == ValueType.none)
            this.element.classList.add('unset-value');
        else
            this.element.classList.remove('unset-value');
        this.disableChangeNotification = true;
        try {
            this.refreshValue(valueType, value);
        }
        catch (err) {
            console.error(err);
        }
        this.disableChangeNotification = false;
    }
}
