// @node-projects/web-component-designer/dist/elements/widgets/demoView/demoView.js
import { BaseCustomWebComponentLazyAppend, css } from '@node-projects/base-custom-webcomponent';
export class DemoView extends BaseCustomWebComponentLazyAppend {
    _placeholder;
    _loading;
    static style = css `
  :host {
    display: block;
    overflow: hidden;
    background: white;
    height: 100%;
    width: 100%;
  }
  #placeholder {
    height: 100%;
    width: 100%;
  }
  #loading {
    position: absolute;
    top: 60px;
    left: 20px;
  }`;
    constructor() {
        super();
        this._placeholder = document.createElement('div');
        this._placeholder.id = 'placeholder';
        this.shadowRoot.appendChild(this._placeholder);
        this._loading = document.createElement('div');
        this._loading.id = 'loading';
        this._loading.textContent = '🛀 Hold on, loading...';
        this.shadowRoot.appendChild(this._loading);
    }
    dispose() {
    }
    executeCommand;
    canExecuteCommand;
    async display(serviceContainer, instanceServiceContainer, code, style) {
        this._loading.hidden = false;
        await serviceContainer.demoProviderService.provideDemo(this._placeholder, serviceContainer, instanceServiceContainer, code, style);
        this._loading.hidden = true;
    }
}
customElements.define('node-projects-demo-view', DemoView);
