// @node-projects/web-component-designer/dist/elements/services/propertiesService/services/AttributesPropertiesService.js
import { ValueType } from '../ValueType.js';
export class AttributesPropertiesService {
    name = "attributes";
    isHandledElement(designItem) {
        return true;
    }
    getProperty(designItem, name) {
        return { name: name, type: 'string', service: this };
    }
    getProperties(designItem) {
        if (designItem) {
            let p = [];
            for (let a of designItem.attributes.keys()) {
                p.push({ name: a, type: 'string', service: this });
            }
            return p;
        }
        return null;
    }
    setValue(designItems, property, value) {
    }
    clearValue(designItems, property) {
    }
    isSet(designItems, property) {
        let all = true;
        let some = false;
        if (designItems != null && designItems.length !== 0) {
            if (designItems.length == 1 && typeof designItems[0].attributes.get(property.name) == 'object')
                return ValueType.bound;
            let attributeName = property.name;
            designItems.forEach((x) => {
                let has = x.attributes.has(attributeName);
                all = all && has;
                some = some || has;
            });
        }
        else
            return ValueType.none;
        return all ? ValueType.all : some ? ValueType.some : ValueType.none;
    }
    getValue(designItems, property) {
        if (designItems != null && designItems.length !== 0) {
            let attributeName = property.name;
            let lastValue = designItems[0].attributes.get(attributeName);
            if (typeof lastValue === 'object')
                return lastValue.rawValue;
            /*
            for (const x of designItems) {
              let value = x.attributes.get(attributeName);
              if (value != lastValue) {
                lastValue = null;
                break;
              }
            }
            */
            return lastValue;
        }
        return null;
    }
    getUnsetValue(designItems, property) {
        return property.defaultValue;
    }
}
