// @node-projects/web-component-designer/dist/elements/widgets/designerView/extensions/grid/DisplayGridExtensionProvider.js
import { css } from "@node-projects/base-custom-webcomponent";
import { DisplayGridExtension } from './DisplayGridExtension.js';
export const gridExtensionShowOverlayOptionName = 'gridExtensionShowOverlay';
export class DisplayGridExtensionProvider {
    gridColor;
    gridFillColor;
    constructor(gridColor = 'orange', gridFillColor = '#ff944722') {
        this.gridColor = gridColor;
        this.gridFillColor = gridFillColor;
    }
    shouldExtend(extensionManager, designerCanvas, designItem) {
        const display = getComputedStyle(designItem.element).display;
        if (display == 'grid' || display == 'inline-grid')
            return designerCanvas.instanceServiceContainer.designContext.extensionOptions[gridExtensionShowOverlayOptionName] !== false;
        return false;
    }
    getExtension(extensionManager, designerCanvas, designItem) {
        return new DisplayGridExtension(extensionManager, designerCanvas, designItem, this.gridColor, this.gridFillColor);
    }
    style = css `
    .svg-grid { stroke: var(--svg-grid-stroke-color); stroke-dasharray: 5; fill: var(--svg-grid-fill-color); }
    .svg-grid-current-cell { stroke: var(--svg-grid-stroke-color); stroke-dasharray: 5; fill: #e3ff4722; }
    .svg-grid-area { font-size: 8px; }
    .svg-grid-gap { stroke: var(--svg-grid-stroke-color); stroke-dasharray: 5; fill: #0000ff22; }
    .svg-grid-header { fill: var(--svg-grid-fill-color); stroke: var(--svg-grid-stroke-color); }
    .svg-grid-plus-sign { stroke: black; }
  `;
}
