// @node-projects/web-component-designer/dist/elements/services/undoService/ChangeGroup.js
export class ChangeGroup {
    title;
    get affectedItems() {
        let s = new Set();
        for (let u of this.undoStack)
            for (let i of u.affectedItems)
                s.add(i);
        for (let u of this.redoStack)
            for (let i of u.affectedItems)
                s.add(i);
        return [...s.values()];
    }
    commitHandler;
    abortHandler;
    constructor(title, commitHandler, abortHandler) {
        this.title = title;
        this.commitHandler = commitHandler;
        this.abortHandler = abortHandler;
    }
    do() {
        let item = null;
        while (item = this.redoStack.pop()) {
            try {
                item.do();
                this.undoStack.push(item);
            }
            catch (err) {
                throw err;
            }
        }
    }
    undo() {
        let item = null;
        while (item = this.undoStack.pop()) {
            try {
                item.undo();
                this.redoStack.push(item);
            }
            catch (err) {
                throw err;
            }
        }
    }
    ;
    commit() {
        this.commitHandler(this);
    }
    abort() {
        this.abortHandler(this);
    }
    mergeWith(other) {
        return false;
    }
    addCommitedSubchangeGroup(changeGroup) {
        this.undoStack.push(changeGroup);
    }
    undoStack = [];
    redoStack = [];
    execute(item) {
        item.do();
        for (let existingItem of this.undoStack) {
            if (existingItem.mergeWith(item))
                return;
        }
        this.undoStack.push(item);
    }
}
