// @node-projects/web-component-designer/dist/elements/services/undoService/transactionItems/AttributeChangeAction.js
export class AttributeChangeAction {
    constructor(designItem, name, newValue, oldValue) {
        this.title = "Change Attribute " + name + " of &lt;" + designItem.name + "&gt;";
        this.designItem = designItem;
        this.name = name;
        this.newValue = newValue;
        this.oldValue = oldValue;
    }
    title;
    get affectedItems() {
        return [this.designItem];
    }
    undo() {
        if (this.oldValue == null) {
            this.designItem._withoutUndoRemoveAttribute(this.name);
        }
        else {
            if (typeof this.oldValue === 'string')
                this.designItem._withoutUndoSetAttribute(this.name, this.oldValue);
            else
                this.designItem._withoutUndoSetAttribute(this.name, this.oldValue.toString());
        }
    }
    do() {
        if (this.newValue == null) {
            this.designItem._withoutUndoRemoveAttribute(this.name);
        }
        else {
            if (typeof this.oldValue === 'string')
                this.designItem._withoutUndoSetAttribute(this.name, this.newValue);
            else
                this.designItem._withoutUndoSetAttribute(this.name, this.newValue.toString());
        }
    }
    designItem;
    name;
    newValue;
    oldValue;
    mergeWith(other) {
        if (other instanceof AttributeChangeAction && this.designItem === other.designItem && this.name === other.name) {
            this.newValue = other.newValue;
            return true;
        }
        return false;
    }
}
