// @node-projects/web-component-designer/dist/elements/services/undoService/transactionItems/AttributeAndPropertyChangeAction.js
export class AttributeAndPropertyChangeAction {
    constructor(designItem, attributeName, propertyName, newValue, oldValue) {
        this.title = "Change Attribute & Property" + attributeName + " of &lt;" + designItem.name + "&gt;";
        this.designItem = designItem;
        this.attributeName = attributeName;
        this.propertyName = propertyName;
        this.newValue = newValue;
        this.oldValue = oldValue;
    }
    title;
    get affectedItems() {
        return [this.designItem];
    }
    undo() {
        this.designItem.element[this.propertyName] = this.oldValue;
        if (this.oldValue == null) {
            this.designItem._withoutUndoRemoveAttribute(this.attributeName);
        }
        else {
            if (typeof this.oldValue === 'string')
                this.designItem._withoutUndoSetAttribute(this.attributeName, this.oldValue);
            else
                this.designItem._withoutUndoSetAttribute(this.attributeName, this.oldValue.toString());
        }
    }
    do() {
        this.designItem.element[this.propertyName] = this.newValue;
        if (this.newValue == null) {
            this.designItem._withoutUndoRemoveAttribute(this.attributeName);
        }
        else {
            if (typeof this.oldValue === 'string')
                this.designItem._withoutUndoSetAttribute(this.attributeName, this.newValue);
            else
                this.designItem._withoutUndoSetAttribute(this.attributeName, this.newValue.toString());
        }
    }
    designItem;
    attributeName;
    propertyName;
    newValue;
    oldValue;
    mergeWith(other) {
        if (other instanceof AttributeAndPropertyChangeAction && this.designItem === other.designItem && this.attributeName === other.attributeName && this.propertyName === other.propertyName) {
            this.newValue = other.newValue;
            return true;
        }
        return false;
    }
}
