// @node-projects/web-component-designer/dist/elements/services/undoService/transactionItems/CssStyleChangeAction copy.js
export class CssStyleChangeAction {
    constructor(designItem, name, newValue, oldValue) {
        this.title = "Change Css Style " + name + " of &lt;" + designItem.name + "&gt;";
        this.designItem = designItem;
        this.name = name;
        this.newValue = newValue;
        this.oldValue = oldValue;
    }
    title;
    get affectedItems() {
        return [this.designItem];
    }
    undo() {
        if (this.oldValue === '' || this.oldValue == null) {
            this.designItem._withoutUndoRemoveStyle(this.name);
            this.designItem.element.style[this.name] = '';
        }
        else {
            this.designItem._withoutUndoSetStyle(this.name, this.oldValue);
            this.designItem.element.style[this.name] = this.oldValue;
        }
    }
    do() {
        if (this.newValue === '' || this.newValue == null) {
            this.designItem._withoutUndoRemoveStyle(this.name);
            this.designItem.element.style[this.name] = '';
        }
        else {
            this.designItem._withoutUndoSetStyle(this.name, this.newValue);
            this.designItem.element.style[this.name] = this.newValue;
        }
    }
    designItem;
    name;
    newValue;
    oldValue;
    mergeWith(other) {
        if (other instanceof CssStyleChangeAction && this.designItem === other.designItem && this.name === other.name) {
            this.newValue = other.newValue;
            return true;
        }
        return false;
    }
}
