// @node-projects/web-component-designer/dist/elements/services/undoService/transactionItems/CssStyleChangeAction.js
export class CssStyleChangeAction {
    constructor(designItem, name, newValue, oldValue) {
        this.title = "Change Css Style " + name + " of &lt;" + designItem.name + "&gt;";
        this.designItem = designItem;
        this.name = name;
        this.newValue = newValue;
        this.oldValue = oldValue;
    }
    title;
    get affectedItems() {
        return [this.designItem];
    }
    undo() {
        if (this.oldValue === '' || this.oldValue == null) {
            this.designItem._withoutUndoRemoveStyle(this.name);
            if (this.name.startsWith('--')) {
                this.designItem.element.style.removeProperty(this.name);
            }
            else {
                this.designItem.element.style[this.name] = '';
            }
            ;
        }
        else {
            this.designItem._withoutUndoSetStyle(this.name, this.oldValue);
            if (this.name.startsWith('--')) {
                this.designItem.element.style.setProperty(this.name, this.oldValue);
            }
            else {
                this.designItem.element.style[this.name] = this.oldValue;
            }
        }
    }
    do() {
        if (this.newValue === '' || this.newValue == null) {
            this.designItem._withoutUndoRemoveStyle(this.name);
            if (this.name.startsWith('--')) {
                this.designItem.element.style.removeProperty(this.name);
            }
            else {
                this.designItem.element.style[this.name] = '';
            }
            ;
        }
        else {
            this.designItem._withoutUndoSetStyle(this.name, this.newValue);
            if (this.name.startsWith('--')) {
                this.designItem.element.style.setProperty(this.name, this.newValue);
            }
            else {
                this.designItem.element.style[this.name] = this.newValue;
            }
        }
    }
    designItem;
    name;
    newValue;
    oldValue;
    mergeWith(other) {
        return false;
    }
}
