// @node-projects/web-component-designer/dist/elements/services/undoService/transactionItems/DeleteAction.js
import { DomHelper } from '@node-projects/base-custom-webcomponent/dist/DomHelper.js';
export class DeleteAction {
    constructor(deletedItems) {
        this.title = "Delete Items";
        this.deletedItems = deletedItems;
    }
    title;
    get affectedItems() {
        return this.deletedItems;
    }
    undo() {
        for (let n = 0; n < this.deletedItems.length; n++) {
            this._parentItems[n]._insertChildInternal(this.deletedItems[n], this._parentIndexes[n]);
        }
        this.affectedItems[0].instanceServiceContainer.contentService.onContentChanged.emit({ changeType: 'added', designItems: this.deletedItems });
    }
    do() {
        this._parentItems = [];
        this._parentIndexes = [];
        for (let n = 0; n < this.deletedItems.length; n++) {
            this._parentItems.push(this.deletedItems[n].parent);
            this._parentIndexes.push(DomHelper.nodeIndex(this.deletedItems[n].element));
        }
        for (let n = 0; n < this.deletedItems.length; n++) {
            this.deletedItems[n].parent._removeChildInternal(this.deletedItems[n]);
        }
        this.affectedItems[0].instanceServiceContainer.contentService.onContentChanged.emit({ changeType: 'removed', designItems: this.deletedItems });
    }
    deletedItems;
    _parentItems;
    _parentIndexes;
    mergeWith(other) {
        return false;
    }
}
