// @node-projects/web-component-designer/dist/elements/services/undoService/transactionItems/InsertChildAction.js
export class InsertChildAction {
    constructor(designItem, newParent, newIndex) {
        this.title = "Move Item";
        this.designItem = designItem;
        this.newParent = newParent;
        this.newIndex = newIndex;
    }
    title;
    get affectedItems() {
        return [this.designItem, this.newParent, this.oldParent];
    }
    undo() {
        this.oldParent._insertChildInternal(this.designItem, this.newIndex);
        this.affectedItems[0].instanceServiceContainer.contentService.onContentChanged.emit({ changeType: 'moved', designItems: [this.designItem] });
    }
    do() {
        this.oldParent = this.designItem.parent;
        this.oldIndex = this.designItem.parent.indexOf(this.designItem);
        this.newParent._insertChildInternal(this.designItem, this.newIndex);
        this.affectedItems[0].instanceServiceContainer.contentService.onContentChanged.emit({ changeType: 'moved', designItems: [this.designItem] });
    }
    designItem;
    newParent;
    newIndex;
    oldParent;
    oldIndex;
    newItem;
    mergeWith(other) {
        return false;
    }
}
