// @node-projects/web-component-designer/dist/elements/services/undoService/transactionItems/InsertAction.js
export class InsertAction {
    constructor(designItem, index, newItem) {
        this.title = "Insert Item";
        this.designItem = designItem;
        this.index = index;
        this.newItem = newItem;
    }
    title;
    get affectedItems() {
        return [this.designItem];
    }
    undo() {
        this.newItem.element.remove();
        this.affectedItems[0].instanceServiceContainer.contentService.onContentChanged.emit({ changeType: 'removed', designItems: [this.newItem] });
    }
    do() {
        this.designItem._insertChildInternal(this.newItem, this.index);
        const prepService = this.designItem.serviceContainer.prepareElementsForDesignerService;
        if (prepService)
            requestAnimationFrame(() => prepService.prepareElementsForDesigner(this.newItem));
        this.affectedItems[0].instanceServiceContainer.contentService.onContentChanged.emit({ changeType: 'added', designItems: [this.newItem] });
    }
    designItem;
    index;
    newItem;
    mergeWith(other) {
        return false;
    }
}
