// @node-projects/web-component-designer/dist/elements/services/undoService/transactionItems/MoveElementInDomAction copy.js
export class MoveElementInDomAction {
    constructor(designItem, newTarget, newPosition, oldTarget, oldPosition) {
        this.title = "Move Item";
        this.designItem = designItem;
        this.newTarget = newTarget;
        this.newPosition = newPosition;
        this.oldTarget = oldTarget;
        this.oldPosition = oldPosition;
    }
    title;
    get affectedItems() {
        return [this.designItem, this.newTarget, this.oldTarget];
    }
    undo() {
        this.oldTarget.element.insertAdjacentElement(this.oldPosition, this.designItem.element);
        this.affectedItems[0].instanceServiceContainer.contentService.onContentChanged.emit({ changeType: 'moved', designItems: [this.designItem] });
    }
    do() {
        //this.newTarget.insertChild(this.designItem, )
        this.newTarget.element.insertAdjacentElement(this.newPosition, this.designItem.element);
        this.affectedItems[0].instanceServiceContainer.contentService.onContentChanged.emit({ changeType: 'moved', designItems: [this.designItem] });
    }
    designItem;
    newTarget;
    newPosition;
    oldTarget;
    oldPosition;
    newItem;
    mergeWith(other) {
        return false;
    }
}
