// @node-projects/web-component-designer/dist/elements/widgets/designerView/extensions/CollaborationCursorOverlayExtension.js
import { AbstractExtension } from './AbstractExtension.js';
import { OverlayLayer } from './OverlayLayer.js';
function truncateText(text, maxLength) {
    if (!text || text.length <= maxLength)
        return text;
    return text.substring(0, Math.max(0, maxLength - 1)) + '…';
}
function createCursorPath(x, y) {
    return [
        `M ${x} ${y}`,
        `L ${x} ${y + 18}`,
        `L ${x + 4.5} ${y + 13.5}`,
        `L ${x + 8.5} ${y + 23}`,
        `L ${x + 12} ${y + 21.5}`,
        `L ${x + 8} ${y + 12.5}`,
        `L ${x + 15} ${y + 12.5}`,
        'Z'
    ].join(' ');
}
export class CollaborationCursorOverlayExtension extends AbstractExtension {
    constructor(extensionManager, designerView, extendedItem) {
        super(extensionManager, designerView, extendedItem);
    }
    extend(cache, event) {
        this.refresh(cache, event);
    }
    refresh(cache, event) {
        const collaborationService = this.extendedItem.instanceServiceContainer.collaborationService;
        if (!collaborationService || !this.extendedItem.isRootItem) {
            if (this.overlays.length)
                this._removeAllOverlays();
            return;
        }
        const localPeerId = collaborationService.session?.peerId;
        const peers = collaborationService.peers
            .filter(peer => peer.peerId !== localPeerId && !!peer.cursorPosition)
            .sort((a, b) => a.peerId.localeCompare(b.peerId));
        const overlayState = JSON.stringify(peers.map(peer => ({
            peerId: peer.peerId,
            displayName: peer.displayName,
            color: peer.color,
            cursorPosition: peer.cursorPosition
        })));
        if (!peers.length) {
            if (this.overlays.length)
                this._removeAllOverlays();
            this._valuesHaveChanges(overlayState);
            return;
        }
        if (!this._valuesHaveChanges(overlayState))
            return;
        this._removeAllOverlays();
        peers.forEach(peer => this.drawPeerCursor(peer));
    }
    dispose() {
        this._removeAllOverlays();
    }
    drawPeerCursor(peer) {
        if (!peer.cursorPosition)
            return;
        const color = peer.color ?? '#1f7ae0';
        const x = peer.cursorPosition.x;
        const y = peer.cursorPosition.y;
        const cursor = this._drawPath(createCursorPath(x, y), 'svg-collaboration-cursor', null, OverlayLayer.Foreground);
        cursor.style.fill = color;
        cursor.style.stroke = '#ffffff';
        cursor.style.strokeWidth = '1.2px';
        this._drawCircle(x, y, 2.5, 'svg-collaboration-cursor-hotspot', null, OverlayLayer.Foreground).style.fill = color;
        this._drawTextWithBackground(truncateText(peer.displayName ?? peer.peerId, 18), x + 16, y + 16, color, 'svg-collaboration-cursor-label', null, OverlayLayer.Foreground);
    }
}
