// @node-projects/web-component-designer/dist/elements/widgets/designerView/extensions/ExtensionType.js
export var ExtensionType;
(function (ExtensionType) {
    ExtensionType[ExtensionType["Permanent"] = 1] = "Permanent";
    ExtensionType[ExtensionType["Selection"] = 2] = "Selection";
    ExtensionType[ExtensionType["PrimarySelection"] = 3] = "PrimarySelection";
    ExtensionType[ExtensionType["PrimarySelectionContainer"] = 4] = "PrimarySelectionContainer";
    ExtensionType[ExtensionType["MouseOver"] = 5] = "MouseOver";
    ExtensionType[ExtensionType["OnlyOneItemSelected"] = 6] = "OnlyOneItemSelected";
    ExtensionType[ExtensionType["MultipleItemsSelected"] = 7] = "MultipleItemsSelected";
    /**
     * Extension for the Container wich the dragged element is draged over.
     */
    ExtensionType[ExtensionType["ContainerDragOver"] = 8] = "ContainerDragOver";
    /**
    * Extension for the Container on wich a new element is draged over.
    */
    ExtensionType[ExtensionType["ContainerExternalDragOver"] = 9] = "ContainerExternalDragOver";
    /**
     * Extension for the Current Container wich the dragged element is contained.
     */
    ExtensionType[ExtensionType["ContainerDrag"] = 10] = "ContainerDrag";
    ExtensionType[ExtensionType["Doubleclick"] = 11] = "Doubleclick";
    ExtensionType[ExtensionType["Placement"] = 12] = "Placement";
})(ExtensionType || (ExtensionType = {}));
