// @node-projects/web-component-designer/dist/elements/widgets/designerView/tools/toolBar/DesignerToolbarButton.js
import { BaseCustomWebComponentConstructorAppend, css, html } from '@node-projects/base-custom-webcomponent';
export class DesignerToolbarButton extends BaseCustomWebComponentConstructorAppend {
    static style = css `
    div {
      width: 24px;
      height: 24px;
      display: flex;
      justify-content: center; 
      align-items: center;
      background-color: inherit;
    }

    div:hover {
      background-color: darkgray;
    }

    img {
      width: calc(100% - 4px);
      height: calc(100% - 4px);
    }    
    `;
    static template = html `<div id="div"><img id="img"></div>`;
    tools;
    popup;
    currentToolOnButton;
    _img;
    _div;
    constructor(designerCanvas, tools) {
        super();
        this.tools = tools;
        this._img = this._getDomElement('img');
        this._div = this._getDomElement('div');
        this._div.onclick = () => {
            if (this.popup) {
                this.getRootNode().host.showPopup(this);
            }
            else {
                this.getRootNode().host.setTool(this.currentToolOnButton);
            }
        };
        this.showTool(Object.getOwnPropertyNames(tools)[0]);
    }
    showTool(name) {
        const tool = this.tools[name];
        if (tool) {
            this._img.title = name;
            this._img.src = tool.icon;
            this.currentToolOnButton = name;
        }
    }
    setActiveTool(name) {
        this.showTool(name);
        const tool = this.tools[name];
        if (tool) {
            this._div.style.backgroundColor = 'lightgreen';
        }
        else {
            this._div.style.backgroundColor = '';
        }
    }
}
customElements.define('node-projects-designer-toolbar-button', DesignerToolbarButton);
