// @node-projects/web-component-designer/dist/elements/services/stylesheetService/AbstractStylesheetService.js
import { TypedEvent } from "@node-projects/base-custom-webcomponent";
import { forceActiveAttributeName, forceFocusAttributeName, forceFocusVisibleAttributeName, forceFocusWithinAttributeName, forceHoverAttributeName, forceVisitedAttributeName } from "../../item/DesignItem.js";
import { calculateSpecificity } from "./SpecificityCalculator.js";
export class AbstractStylesheetService {
    _stylesheets = new Map();
    _documentStylesheets = new Map();
    _allStylesheets = new Map();
    _instanceServiceContainer;
    constructor(designerCanvas) {
        this._instanceServiceContainer = designerCanvas.instanceServiceContainer;
    }
    async setStylesheets(stylesheets) {
        await this.internalSetStylesheets(stylesheets, this._stylesheets);
    }
    async setDocumentStylesheets(stylesheets) {
        await this.internalSetStylesheets(stylesheets, this._documentStylesheets);
    }
    async internalSetStylesheets(stylesheets, targetMap) {
        if (targetMap != null && stylesheets != null && targetMap.size == stylesheets.length && stylesheets.every(x => targetMap.has(x.name))) {
            for (let stylesheet of stylesheets) {
                const old = targetMap.get(stylesheet.name);
                if (old.stylesheet.content != stylesheet.content) {
                    try {
                        targetMap.set(stylesheet.name, {
                            stylesheet: stylesheet,
                            ast: await this.internalParse(stylesheet.content)
                        });
                    }
                    catch (err) {
                        console.warn("error parsing stylesheet", stylesheet, err);
                    }
                    if (targetMap == this._stylesheets)
                        this.stylesheetChanged.emit({ name: stylesheet.name, newStyle: stylesheet.content, oldStyle: old.stylesheet.content, changeSource: 'extern' });
                }
            }
        }
        else if (stylesheets != null) {
            targetMap.clear();
            for (let stylesheet of stylesheets) {
                let ast = null;
                try {
                    ast = await this.internalParse(stylesheet.content);
                }
                catch (err) {
                    console.warn("error parsing stylesheet", stylesheet, err);
                }
                targetMap.set(stylesheet.name, {
                    stylesheet: stylesheet,
                    ast: ast
                });
            }
            if (targetMap == this._stylesheets)
                this.stylesheetsChanged.emit();
        }
        else {
            targetMap.clear();
        }
        this._allStylesheets.clear();
        for (let s of this._documentStylesheets) {
            this._allStylesheets.set(s[0], s[1]);
        }
        for (let s of this._stylesheets) {
            this._allStylesheets.set(s[0], s[1]);
        }
    }
    async internalReparseStylesheet(name) {
        let lst = this._allStylesheets;
        if (this._documentStylesheets.has(name))
            lst = this._documentStylesheets;
        if (this._stylesheets.has(name))
            lst = this._stylesheets;
        const ss = lst.get(name);
        let ast = null;
        try {
            ast = await this.internalParse(ss.stylesheet.content);
        }
        catch (err) {
            console.warn("error parsing stylesheet", name, err);
        }
        const v = {
            stylesheet: ss.stylesheet,
            ast: ast
        };
        this._stylesheets.set(name, v);
        this._allStylesheets.set(name, v);
    }
    //TODO: rename to externalStylesheets
    getStylesheets() {
        let stylesheets = [];
        for (let item of this._stylesheets) {
            stylesheets.push(item[1].stylesheet);
        }
        ;
        /*for (let item of this._documentStylesheets) {
            stylesheets.push(item[1].stylesheet);
        };*/
        return stylesheets;
    }
    updateDeclarationValue(declaration, value, important) {
        this.updateDeclarationValueWithoutUndo(declaration, value, important);
    }
    stylesheetChanged = new TypedEvent();
    stylesheetsChanged = new TypedEvent();
    static patchStylesheetSelectorForDesigner(text) {
        return text.replaceAll(':hover', '[' + forceHoverAttributeName + ']')
            .replaceAll(':active', '[' + forceActiveAttributeName + ']')
            .replaceAll(':visited', '[' + forceVisitedAttributeName + ']')
            .replaceAll(':focus', '[' + forceFocusAttributeName + ']')
            .replaceAll(':focus-within', '[' + forceFocusWithinAttributeName + ']')
            .replaceAll(':focus-visible', '[' + forceFocusVisibleAttributeName + ']');
    }
    elementMatchesASelector(designItem, selectors) {
        if (designItem == null)
            return false;
        let s = null;
        for (let selector of selectors) {
            const patched = AbstractStylesheetService.patchStylesheetSelectorForDesigner(selector);
            try {
                if (designItem.element.matches(patched)) {
                    let spec = calculateSpecificity(selector);
                    if (s === null || spec.A > s.A || spec.B > s.B || spec.C > s.C)
                        s = spec;
                }
            }
            catch (err) {
                console.warn("invalid selector: ", selector, "patched: " + patched);
            }
        }
        return s === null ? false : s;
    }
}
