// @node-projects/web-component-designer/dist/elements/widgets/designerView/extensions/ExtensionType.js
export var ExtensionType;
(function (ExtensionType) {
    ExtensionType[ExtensionType["Permanent"] = 1] = "Permanent";
    ExtensionType[ExtensionType["Selection"] = 2] = "Selection";
    ExtensionType[ExtensionType["PrimarySelection"] = 4] = "PrimarySelection";
    ExtensionType[ExtensionType["PrimarySelectionContainer"] = 8] = "PrimarySelectionContainer";
    ExtensionType[ExtensionType["MouseOver"] = 16] = "MouseOver";
    ExtensionType[ExtensionType["OnlyOneItemSelected"] = 32] = "OnlyOneItemSelected";
    ExtensionType[ExtensionType["MultipleItemsSelected"] = 64] = "MultipleItemsSelected";
    /**
     * Extension for the Container wich the dragged element is draged over.
     */
    ExtensionType[ExtensionType["ContainerDragOver"] = 128] = "ContainerDragOver";
    /**
     * Extension for the Current Container wich the dragged element is contained.
     */
    ExtensionType[ExtensionType["ContainerDrag"] = 256] = "ContainerDrag";
    ExtensionType[ExtensionType["Doubleclick"] = 512] = "Doubleclick";
})(ExtensionType || (ExtensionType = {}));
