// @node-projects/web-component-designer/dist/elements/controls/ImageButtonListSelector.js
import { BaseCustomWebComponentConstructorAppend, css, html, TypedEvent } from '@node-projects/base-custom-webcomponent';
export class ImageButtonListSelector extends BaseCustomWebComponentConstructorAppend {
    static style = css `
    div {
      font-size: 10px;
      color: white;
    }
    #property {
      color: #00aff0;
    }
    #value {
      color: lightgray;
    }
    #value.value-set {
      color: wheat;
    }
    .container {
      display: flex;
      flex-wrap: wrap;
      flex-direction: row;
    }
    ::slotted(button) {
      min-width: 24px;
      height: 24px;
      padding: 1px;
      background: white;
      border: 1px solid lightgray;
    }
  `;
    static template = html `
    <div>
      <div><span id="property"></span>: <span id="value"></span></div>
      <div class="container"><slot id="slot"></slot></div>
    </div>
  `;
    _value;
    get value() {
        return this._value;
    }
    set value(value) {
        const oldValue = this._value;
        this._value = value;
        this._updateValue();
        this.valueChanged.emit({ newValue: this._value, oldValue: oldValue });
    }
    valueChanged = new TypedEvent();
    property;
    unsetValue;
    _updateValue() {
        if (this.value) {
            this._getDomElement('value').innerText = this.value;
            this._getDomElement('value').classList.add('value-set');
        }
        else {
            this._getDomElement('value').classList.remove('value-set');
        }
        const slot = this._getDomElement('slot');
        for (let e of slot.assignedElements()) {
            if (e.dataset.value == this.value)
                e.style.background = "cornflowerblue";
            else
                e.style.background = "";
        }
    }
    ready() {
        this._parseAttributesToProperties();
        const slot = this._getDomElement('slot');
        slot.onclick = (e) => {
            const path = e.composedPath();
            for (let e of slot.assignedElements()) {
                if (path.indexOf(e) >= 0) {
                    this.value = e.dataset.value;
                }
            }
        };
        this._getDomElement('property').innerText = this.property ?? '';
        this._getDomElement('value').innerText = this.unsetValue ?? '';
        this._updateValue();
    }
}
customElements.define('node-projects-image-button-list-selector', ImageButtonListSelector);
