// @node-projects/web-component-designer/dist/commandHandling/CommandType.js
export var CommandType;
(function (CommandType) {
    CommandType["copy"] = "copy";
    CommandType["paste"] = "paste";
    CommandType["cut"] = "cut";
    CommandType["delete"] = "delete";
    CommandType["undo"] = "undo";
    CommandType["redo"] = "redo";
    CommandType["holdUndo"] = "holdUndo";
    CommandType["holdRedo"] = "holdRedo";
    CommandType["rotateCounterClockwise"] = "rotateCounterClockwise";
    CommandType["rotateClockwise"] = "rotateClockwise";
    CommandType["selectAll"] = "selectAll";
    CommandType["moveToFront"] = "moveToFront";
    CommandType["moveForward"] = "moveForward";
    CommandType["moveBackward"] = "moveBackward";
    CommandType["moveToBack"] = "moveToBack";
    CommandType["arrangeLeft"] = "arrangeLeft";
    CommandType["arrangeCenter"] = "arrangeCenter";
    CommandType["arrangeRight"] = "arrangeRight";
    CommandType["arrangeTop"] = "arrangeTop";
    CommandType["arrangeMiddle"] = "arrangeMiddle";
    CommandType["arrangeBottom"] = "arrangeBottom";
    CommandType["unifyWidth"] = "unifyWidth";
    CommandType["unifyHeight"] = "unifyHeight";
    CommandType["distributeHorizontal"] = "distributeHorizontaly";
    CommandType["distributeVertical"] = "distributeVertical";
    CommandType["setTool"] = "setTool";
    CommandType["setStrokeColor"] = "setStrokeColor";
    CommandType["setFillBrush"] = "setFillBrush";
    CommandType["setStrokeThickness"] = "setStrokeThickness";
    CommandType["screenshot"] = "screenshot";
})(CommandType || (CommandType = {}));
