// @node-projects/base-custom-webcomponent/dist/BaseCustomWebComponent.js
import { addTouchFriendlyContextMenu } from "./TouchContextMenu.js";
import { TypedEvent } from './TypedEvent.js';
export const html = function (strings, ...values) {
    const template = document.createElement('template');
    template.innerHTML = strings.raw[0];
    return template;
};
export const htmlFromString = function (value) {
    const template = document.createElement('template');
    template.innerHTML = value;
    return template;
};
export const css = function (strings, ...values) {
    const cssStyleSheet = new CSSStyleSheet();
    //@ts-ignore
    cssStyleSheet.replaceSync(strings.raw[0]);
    return cssStyleSheet;
};
export const cssFromString = function (value) {
    const cssStyleSheet = new CSSStyleSheet();
    //@ts-ignore
    cssStyleSheet.replaceSync(value);
    return cssStyleSheet;
};
export const cssAsync = async function (strings, ...values) {
    const cssStyleSheet = new CSSStyleSheet();
    //@ts-ignore
    await cssStyleSheet.replace(strings.raw[0]);
    return cssStyleSheet;
};
// decorators
export function property(par) {
    return function (target, propertyKey) {
        //@ts-ignore
        if (!target.constructor.properties) {
            //@ts-ignore
            target.constructor.properties = {};
        }
        if (par && par.type != null) {
            //@ts-ignore
            target.constructor.properties[propertyKey] = par.type ? par.type : String;
        }
        else {
            //@ts-ignore
            target.constructor.properties[propertyKey] = par ? par : String;
        }
    };
}
export function customElement(tagname) {
    return function (class_) {
        //@ts-ignore
        class_.is = tagname;
        customElements.define(tagname, class_);
    };
}
export class BaseCustomWebComponentNoAttachedTemplate extends HTMLElement {
    static style;
    static template;
    _bindings;
    _repeatWeakMap;
    //@ts-ignore
    static _bindingRegex = /\[\[.*?\]\]/g;
    _getDomElement(id) {
        if (this.shadowRoot.children.length > 1 || (this.shadowRoot.children[0] !== undefined && this.shadowRoot.children[0].localName !== 'style'))
            return this.shadowRoot.getElementById(id);
        return this._rootDocumentFragment.getElementById(id);
    }
    _getDomElements(selector) {
        if (this.shadowRoot.children.length > 1 || (this.shadowRoot.children[0] !== undefined && this.shadowRoot.children[0].localName !== 'style'))
            return this.shadowRoot.querySelectorAll(selector);
        return this._rootDocumentFragment.querySelectorAll(selector);
    }
    _assignEvents(node) {
        if (!node) {
            node = this.shadowRoot.children.length > 0 ? this.shadowRoot : this._rootDocumentFragment;
        }
        if (node instanceof Element) {
            for (let a of node.attributes) {
                if (a.name.startsWith('@') && !a.value.startsWith('[[')) {
                    try {
                        if (a.name == "@touch:contextmenu")
                            addTouchFriendlyContextMenu(node, this[a.value].bind(this));
                        else {
                            const sNm = a.name.substr(1);
                            let nm = sNm.replace(/-([a-z])/g, (g) => g[1].toUpperCase());
                            if (node[nm] instanceof TypedEvent) {
                                node[nm].on(this[a.value].bind(this));
                            }
                            else {
                                node.addEventListener(sNm, this[a.value].bind(this));
                            }
                        }
                    }
                    catch (error) {
                        console.warn(error.message, 'Failed to attach event "', a, node);
                    }
                }
            }
        }
        for (let n of node.childNodes) {
            this._assignEvents(n);
        }
    }
    /**
     * Parses Polymer like Bindings
     *
     * use [[expression]] for one way bindings
     *
     * use {{this.property::change;paste}} for two way wich binds to events 'change 'and 'paste'
     *
     * use @eventname="eventHandler" to bind a handler to a event
     * or @eventname="[[this.eventHandler(par1, par2, ..)]]" for complexer event logic
     * use @touch:contextmenu... for a context menu that also works with long press on touch
     *
     * use css:cssPropertyName=[[expression]] to bind to a css property
     *
     * use class:className=[[boolExpression]] to set/remove a css class
     *
     * sub <template></template> elements are not bound, so elemnts like <iron-list> of polymer also work
     *
     * use repeat:nameOfItem=[[enumerableExpression]] on a Template Element to repeate it for every instance of the enumarable
     * ==> this could also be nested
     *
     */
    _bindingsParse(node, removeAttributes = false, host = null, context = null) {
        this._bindingsInternalParse(node, null, removeAttributes, host, context);
    }
    _bindingsInternalParse(node, repeatBindingItems, removeAttributes, host, context) {
        if (!this._bindings)
            this._bindings = [];
        if (!node)
            node = this.shadowRoot.childNodes.length > 0 ? this.shadowRoot : this._rootDocumentFragment;
        if (node instanceof Element) { //node.nodeType === 1
            const attributes = Array.from(node.attributes);
            for (let a of attributes) {
                if (a.name.startsWith('css:') && a.value.startsWith('[[') && a.value.endsWith(']]')) {
                    const value = a.value.substring(2, a.value.length - 2).replaceAll('&amp;', '&');
                    const camelCased = a.name.substring(4, a.name.length).replace(/-([a-z])/g, (g) => g[1].toUpperCase());
                    this._bindings.push(() => this._bindingSetElementCssValue(node, camelCased, value, repeatBindingItems, host, context));
                    this._bindings[this._bindings.length - 1](true);
                }
                else if (a.name.startsWith('class:') && a.value.startsWith('[[') && a.value.endsWith(']]')) {
                    const value = a.value.substring(2, a.value.length - 2).replaceAll('&amp;', '&');
                    const camelCased = a.name.substring(6, a.name.length).replace(/-([a-z])/g, (g) => g[1].toUpperCase());
                    this._bindings.push(() => this._bindingSetElementClass(node, camelCased, value, repeatBindingItems, host, context));
                    this._bindings[this._bindings.length - 1](true);
                }
                else if (a.name == 'repeat-changed-item-callback') {
                    //do nothing
                }
                else if (a.name.startsWith('repeat:') && a.value.startsWith('[[') && a.value.endsWith(']]')) {
                    const value = a.value.substring(2, a.value.length - 2).replaceAll('&amp;', '&');
                    const bindingItemVariableName = a.name.substring(7, a.name.length).replace(/-([a-z])/g, (g) => g[1].toUpperCase());
                    const elementsCache = [];
                    let bindingIndexname = 'index';
                    let changeItemCallback = null;
                    const indexNameAttribute = attributes.find(x => x.name == 'repeat-index');
                    if (indexNameAttribute)
                        bindingIndexname = indexNameAttribute.value;
                    const changeItemCallbackAttribute = attributes.find(x => x.name == 'repeat-changed-item-callback');
                    if (changeItemCallbackAttribute)
                        changeItemCallback = changeItemCallbackAttribute.value;
                    this._bindings.push(() => this._bindingRepeat(node, bindingItemVariableName, bindingIndexname, value, changeItemCallback, repeatBindingItems, elementsCache, host, context));
                    this._bindings[this._bindings.length - 1](true);
                }
                else if (a.name.startsWith('@') && a.value.startsWith('[[') && a.value.endsWith(']]')) { //todo remove events on repeat refresh
                    const value = a.value.substring(2, a.value.length - 2).replaceAll('&amp;', '&');
                    const camelCased = a.name.substring(1, a.name.length).replace(/-([a-z])/g, (g) => g[1].toUpperCase());
                    if (a.name == "@touch:contextmenu")
                        addTouchFriendlyContextMenu(node, (e) => this._bindingRunEval(value, repeatBindingItems, e, host, context));
                    else {
                        if (node[camelCased] instanceof TypedEvent) {
                            node[camelCased].on((e) => this._bindingRunEval(value, repeatBindingItems, e, host, context));
                        }
                        else {
                            node.addEventListener(camelCased, (e) => this._bindingRunEval(value, repeatBindingItems, e, host, context));
                        }
                    }
                }
                else if (a.value.startsWith('[[') && a.value.endsWith(']]')) {
                    let value = a.value.substring(2, a.value.length - 2).replaceAll('&amp;', '&');
                    const camelCased = a.name.replace(/-([a-z])/g, (g) => g[1].toUpperCase());
                    let noNull = false;
                    if (value.startsWith('?')) {
                        value = value.substring(1);
                        noNull = true;
                    }
                    this._bindings.push((firstRun) => this._bindingSetNodeValue(firstRun, node, a, camelCased, value, repeatBindingItems, removeAttributes, host, context, noNull));
                    this._bindings[this._bindings.length - 1](true);
                }
                else if (a.value.startsWith('{{') && a.value.endsWith('}}')) {
                    const attributeValues = a.value.substring(2, a.value.length - 2).split('::');
                    let value = attributeValues[0];
                    let event = 'input';
                    if (attributeValues.length > 1 && attributeValues[1])
                        event = attributeValues[1];
                    const camelCased = a.name.replace(/-([a-z])/g, (g) => g[1].toUpperCase());
                    let noNull = false;
                    if (value.startsWith('?')) {
                        value = value.substring(1);
                        noNull = true;
                    }
                    this._bindings.push((firstRun) => this._bindingSetNodeValue(firstRun, node, a, camelCased, value, repeatBindingItems, removeAttributes, host, context, noNull));
                    this._bindings[this._bindings.length - 1](true);
                    if (event) {
                        event.split(';').forEach(x => node.addEventListener(x, (e) => this._bindingsSetValue(this, value, node[camelCased])));
                    }
                }
            }
        }
        else if (node.nodeType === 3) {
            if (node.nodeValue.indexOf('[[') >= 0) {
                const text = node.nodeValue;
                const matches = text.matchAll(this.constructor._bindingRegex);
                let lastindex = 0;
                let fragment;
                for (let m of matches) {
                    if (m.index == 0 && m[0].length == text.length && node.parentNode.childNodes.length == 1) {
                        const value = m[0].substr(2, m[0].length - 4);
                        const parent = node.parentNode;
                        node.parentNode.removeChild(node);
                        this._bindings.push((firstRun) => this._bindingSetNodeValue(firstRun, parent, null, 'innerHTML', value, repeatBindingItems, removeAttributes, host, context, false));
                        this._bindings[this._bindings.length - 1](true);
                    }
                    else {
                        if (!fragment)
                            fragment = document.createDocumentFragment();
                        if (m.index - lastindex > 0) {
                            const tn = document.createTextNode(text.substr(lastindex, m.index - lastindex));
                            fragment.appendChild(tn);
                        }
                        const newNode = document.createElement('span');
                        const value = m[0].substr(2, m[0].length - 4);
                        this._bindings.push((firstRun) => this._bindingSetNodeValue(firstRun, newNode, null, 'innerHTML', value, repeatBindingItems, removeAttributes, host, context, false));
                        this._bindings[this._bindings.length - 1](true);
                        fragment.appendChild(newNode);
                        lastindex = m.index + m[0].length;
                    }
                }
                if (fragment) {
                    if (lastindex > 0 && text.length - lastindex > 0) {
                        let tn = document.createTextNode(text.substr(lastindex, text.length - lastindex));
                        fragment.appendChild(tn);
                    }
                    node.parentNode.replaceChild(fragment, node);
                }
            }
        }
        if (!(node instanceof HTMLTemplateElement)) {
            let children = Array.from(node.childNodes);
            for (let n of children) {
                this._bindingsInternalParse(n, repeatBindingItems, removeAttributes, host, context);
            }
        }
    }
    _bindingRunEval(expression, repeatBindingItems, event, host, context) {
        if (host)
            return this._bindingRunEvalInt.bind(host)(expression, repeatBindingItems, event, context);
        return this._bindingRunEvalInt(expression, repeatBindingItems, event, context);
    }
    //This method can not use "this" anywhere, cause it's bound to different host via method above.
    _bindingRunEvalInt(expression, repeatBindingItems, event, context) {
        if (repeatBindingItems) {
            let n = 0;
            let set = new Set();
            for (let b of repeatBindingItems) {
                if (!set.has(b.name)) {
                    expression = 'let ' + b.name + ' = ___repeatBindingItems[' + n + '].item;' + expression;
                    set.add(b.name);
                }
                n++;
            }
            if (event) {
                expression = 'let event = ___event;' + expression;
            }
            if (context) {
                for (let i in context) {
                    expression = 'let ' + i + ' = ___context["' + i + '"];' + expression;
                }
            }
            //@ts-ignore
            var ___repeatBindingItems = repeatBindingItems;
            //@ts-ignore
            var ___event = event;
            //@ts-ignore
            var ___context = context;
            let value = eval(expression);
            return value;
        }
        if (context) {
            for (let i in context) {
                expression = 'let ' + i + ' = ___context["' + i + '"];' + expression;
            }
            //@ts-ignore
            var ___context = context;
            let value = eval(expression);
            return value;
        }
        let value = eval(expression);
        return value;
    }
    _bindingRepeat(node, bindingProperty, bindingIndexName, expression, callback, repeatBindingItems, elementsCache, host, context) {
        try {
            const values = this._bindingRunEval(expression, repeatBindingItems, null, host, context);
            if (values) {
                if (callback) {
                    if (callback.startsWith('[[') && callback.endsWith(']]'))
                        callback = callback.substring(2, callback.length - 2);
                    else
                        callback = "this." + callback;
                }
                for (let c of elementsCache) { // todo bindings of childs need to be killed
                    if (c.parentNode) {
                        let intRepeatBindingItems = [];
                        intRepeatBindingItems.push({ name: 'nodes', item: [c] });
                        intRepeatBindingItems.push({ name: 'callbackType', item: 'remove' });
                        this._bindingRunEval(callback, intRepeatBindingItems, null, host, context);
                        c.parentNode.removeChild(c);
                    }
                }
                let i = 0;
                for (let e of values) {
                    let intRepeatBindingItems = [];
                    if (repeatBindingItems)
                        intRepeatBindingItems = repeatBindingItems.slice();
                    intRepeatBindingItems.push({ name: bindingProperty, item: e });
                    intRepeatBindingItems.push({ name: bindingIndexName, item: i });
                    let nd = node.content.cloneNode(true);
                    elementsCache.push(...nd.children);
                    this._bindingsInternalParse(nd, intRepeatBindingItems, true, host, context);
                    if (callback) {
                        intRepeatBindingItems.push({ name: 'nodes', item: nd.children });
                        intRepeatBindingItems.push({ name: 'callbackType', item: 'create' });
                        let nds = this._bindingRunEval(callback, intRepeatBindingItems, null, host, context);
                        if (nds === undefined)
                            nds = nd.children;
                        if (nds) {
                            for (let n of Array.from(nds))
                                node.parentNode.appendChild(n);
                        }
                    }
                    else {
                        node.parentNode.appendChild(nd);
                    }
                    i++;
                }
            }
        }
        catch (error) {
            console.warn(error.message, 'Failed to bind Repeat "' + bindingProperty + '" to expression "' + expression + '"', node);
        }
    }
    _bindingSetNodeValue(firstRun, node, attribute, property, expression, repeatBindingItems, removeAttributes, host, context, noNull) {
        try {
            const value = this._bindingRunEval(expression, repeatBindingItems, null, host, context);
            if (firstRun || node[property] !== value) {
                if (removeAttributes && attribute)
                    node.removeAttribute(attribute.name);
                if (property === 'innerHTML' && value instanceof Element) {
                    for (let c = node.firstChild; c !== null; c = node.firstChild) {
                        node.removeChild(c);
                    }
                    node.appendChild(value);
                }
                else {
                    if (property[0] == '$') {
                        if (!value && noNull)
                            node.setAttribute(property.substring(1, property.length), '');
                        else
                            node.setAttribute(property.substring(1, property.length), value);
                    }
                    else if (property == 'class')
                        node.setAttribute(property, value);
                    else {
                        if (!value && noNull)
                            node[property] = '';
                        else
                            node[property] = value;
                    }
                }
            }
        }
        catch (error) {
            console.warn(error.message, 'Failed to bind Property "' + property + '" to expression "' + expression + '"', node);
        }
    }
    _bindingSetElementCssValue(node, property, expression, repeatBindingItems, host, context) {
        try {
            const value = this._bindingRunEval(expression, repeatBindingItems, null, host, context);
            if (node.style[property] !== value)
                node.style[property] = value;
        }
        catch (error) {
            console.warn(error.message, 'Failed to bind CSS Property "' + property + '" to expression "' + expression + '"', node);
        }
    }
    _bindingSetElementClass(node, classname, expression, repeatBindingItems, host, context) {
        try {
            const value = this._bindingRunEval(expression, repeatBindingItems, null, host, context);
            if (value) {
                if (!node.classList.contains(classname))
                    node.classList.add(classname);
            }
            else {
                if (node.classList.contains(classname))
                    node.classList.remove(classname);
            }
        }
        catch (error) {
            console.warn(error.message, 'Failed to bind CSS Class "' + classname + '" to expression "' + expression + '"', node);
        }
    }
    _bindingsRefresh(property) {
        if (this._bindings)
            this._bindings.forEach(x => x(false));
    }
    _bindingsSetValue(obj, path, value) {
        if (path === undefined || path === null) {
            return;
        }
        if (path.startsWith('this.')) {
            path = path.substr(5);
        }
        const pathParts = path.split('.');
        for (let i = 0; i < pathParts.length - 1; i++) {
            if (obj != null) {
                let newObj = obj[pathParts[i]];
                if (newObj == null) {
                    newObj = {};
                    obj[pathParts[i]] = newObj;
                }
                obj = newObj;
            }
        }
        obj[pathParts[pathParts.length - 1]] = value;
    }
    //@ts-ignore
    static _propertiesDictionary;
    _parseAttributesToProperties() {
        //@ts-ignore
        if (!this.constructor._propertiesDictionary) {
            //@ts-ignore
            this.constructor._propertiesDictionary = new Map();
            //@ts-ignore
            for (let i in this.constructor.properties) {
                //@ts-ignore
                this.constructor._propertiesDictionary.set(i.replace(/([A-Z])/g, (g) => `-${g[0].toLowerCase()}`), [i, this.constructor.properties[i]]);
            }
        }
        for (const a of this.attributes) {
            //@ts-ignore
            let pair = this.constructor._propertiesDictionary.get(a.name);
            if (pair) {
                if (pair[1] === Boolean)
                    this[pair[0]] = true;
                else if (pair[1] === Object) {
                    if (!a.value.startsWith("{{") && !a.value.startsWith("[["))
                        this[pair[0]] = JSON.parse(a.value);
                }
                else
                    this[pair[0]] = a.value;
            }
        }
    }
    /*attributeChangedCallback(name, oldValue, newValue) {
      //@ts-ignore
      if (this.constructor._propertiesDictionary) {
        this._parseAttributesToProperties();
      }
    }*/
    async _waitForChildrenReady() {
        await Promise.all(Array.from(this.shadowRoot.querySelectorAll(':not(:defined)'), n => customElements.whenDefined(n.localName)));
    }
    _rootDocumentFragment;
    constructor(template, style) {
        super();
        this.attachShadow({ mode: 'open' });
        if (template) {
            //@ts-ignore
            this._rootDocumentFragment = template.content.cloneNode(true);
        }
        //@ts-ignore
        else if (this.constructor.template) {
            //@ts-ignore
            this._rootDocumentFragment = this.constructor.template.content.cloneNode(true);
        }
        if (style) {
            //@ts-ignore
            if (style instanceof Promise)
                //@ts-ignore
                style.then((s) => this.shadowRoot.adoptedStyleSheets = [s]);
            else
                //@ts-ignore
                this.shadowRoot.adoptedStyleSheets = [style];
            //@ts-ignore
        }
        else if (this.constructor.style) {
            //@ts-ignore
            if (this.constructor.style instanceof Promise)
                //@ts-ignore
                this.constructor.style.then((style) => this.shadowRoot.adoptedStyleSheets = [style]);
            else
                //@ts-ignore
                this.shadowRoot.adoptedStyleSheets = [this.constructor.style];
        }
    }
}
export class BaseCustomWebComponentLazyAppend extends BaseCustomWebComponentNoAttachedTemplate {
    constructor(template, style) {
        super(template, style);
        queueMicrotask(() => {
            if (this._rootDocumentFragment)
                this.shadowRoot.appendChild(this._rootDocumentFragment);
            //@ts-ignore
            if (this.oneTimeSetup && !this.constructor._oneTimeSetup) {
                //@ts-ignore
                this.constructor._oneTimeSetup = true;
                //@ts-ignore
                this.oneTimeSetup();
            }
            //@ts-ignore
            if (this.ready)
                //@ts-ignore
                this.ready();
        });
    }
}
export class BaseCustomWebComponentConstructorAppend extends BaseCustomWebComponentNoAttachedTemplate {
    constructor(template, style) {
        super(template, style);
        if (this._rootDocumentFragment)
            this.shadowRoot.appendChild(this._rootDocumentFragment);
        queueMicrotask(() => {
            //@ts-ignore
            if (this.oneTimeSetup && !this.constructor._oneTimeSetup) {
                //@ts-ignore
                this.constructor._oneTimeSetup = true;
                //@ts-ignore
                this.oneTimeSetup();
            }
            //@ts-ignore
            if (this.ready)
                //@ts-ignore
                this.ready();
        });
    }
}
export class BaseCustomWebComponentConstructorAppendLazyReady extends BaseCustomWebComponentNoAttachedTemplate {
    constructor(template, style) {
        super(template, style);
        if (this._rootDocumentFragment)
            this.shadowRoot.appendChild(this._rootDocumentFragment);
        requestAnimationFrame(() => {
            //@ts-ignore
            if (this.oneTimeSetup && !this.constructor._oneTimeSetup) {
                //@ts-ignore
                this.constructor._oneTimeSetup = true;
                //@ts-ignore
                this.oneTimeSetup();
            }
            //@ts-ignore
            if (this.ready)
                //@ts-ignore
                this.ready();
        });
    }
}
