// @node-projects/base-custom-webcomponent/dist/TypedEvent.js
export class PropertyChangedArgs {
    constructor(newValue, oldValue) {
        this.newValue = newValue;
        this.oldValue = oldValue;
    }
}
export class TypedEvent {
    constructor() {
        this.listeners = [];
        this.listenersOncer = [];
        this.listenerSingle = null;
        this.on = (listener) => {
            this.listeners.push(listener);
            return {
                dispose: () => this.off(listener)
            };
        };
        this.single = (listener) => {
            this.listenerSingle = listener;
            return {
                dispose: () => { if (this.listenerSingle === listener)
                    this.listenerSingle = null; }
            };
        };
        this.once = (listener) => {
            this.listenersOncer.push(listener);
        };
        this.off = (listener) => {
            let callbackIndex = this.listeners.indexOf(listener);
            if (callbackIndex > -1)
                this.listeners.splice(callbackIndex, 1);
        };
        this.emit = (event) => {
            this.listeners.forEach((listener) => listener(event));
            if (this.listenersOncer.length > 0) {
                const toCall = this.listenersOncer;
                this.listenersOncer = [];
                toCall.forEach(listener => listener(event));
            }
            if (this.listenerSingle) {
                this.listenerSingle(event);
            }
        };
        this.pipe = (te) => {
            return this.on((e) => te.emit(e));
        };
    }
}
