// @node-projects/web-component-designer/dist/elements/widgets/designerView/tools/NamedTools.js
export var NamedTools;
(function (NamedTools) {
    NamedTools["Pointer"] = "Pointer";
    NamedTools["DrawSelection"] = "DrawSelection";
    NamedTools["DrawPath"] = "DrawPath";
    NamedTools["DrawRect"] = "DrawRect";
    NamedTools["DrawEllipsis"] = "DrawEllipsis";
    NamedTools["DrawLine"] = "DrawLine";
    NamedTools["Zoom"] = "Zoom";
    NamedTools["Pan"] = "Pan";
    NamedTools["MagicWandSelector"] = "MagicWandSelector";
    NamedTools["RectangleSelector"] = "RectangleSelector";
    NamedTools["PickColor"] = "PickColor";
    NamedTools["Text"] = "Text";
    NamedTools["DrawElementTool"] = "DrawElementTool";
    NamedTools["Guides"] = "Guides";
    NamedTools["Inspect"] = "Inspect";
    NamedTools["Accessibility"] = "Accessibility";
    NamedTools["Position"] = "Position";
    NamedTools["Padding"] = "Padding";
    NamedTools["Margin"] = "Margin";
    NamedTools["Flexbox"] = "Flexbox";
    NamedTools["Move"] = "Move";
    NamedTools["HueShift"] = "HueShift";
    NamedTools["BoxShadows"] = "BoxShadows";
    NamedTools["FontStyles"] = "FontStyles";
})(NamedTools || (NamedTools = {}));
