// @node-projects/web-component-designer/dist/elements/widgets/paletteView/paletteTreeView.js
import { css, html, BaseCustomWebComponentConstructorAppend } from '/webui/node_modules/@node-projects/base-custom-webcomponent/./dist/index.js';
import { dragDropFormatNameElementDefinition } from '../../../Constants.js';
export class PaletteTreeView extends BaseCustomWebComponentConstructorAppend {
    _treeDiv;
    _tree;
    _filter;
    static style = css `
      * {
          touch-action: none;
      }

      span.drag-source {
        border: 1px solid grey;
        border-radius: 3px;
        padding: 2px;
        background-color: silver;
      }

      span.fancytree-node.fancytree-drag-source {
        outline: 1px dotted grey;
      }
      span.fancytree-node.fancytree-drop-accept {
        outline: 1px dotted green;
      }
      span.fancytree-node.fancytree-drop-reject {
        outline: 1px dotted red;
      }
      #tree ul {
        border: none;
      }
      #tree ul:focus {
        outline: none;
      }
      span.fancytree-title {
        align-items: center;
        flex-direction: row;
        display: inline-flex;
      }
      td {
        white-space: nowrap;
      }
      td:nth-child(n+2) {
        text-align: center;
      }
      td > img {
        vertical-align: middle;
      }
    `;
    static template = html `
  <div style="height: 100%;">
    <input id="input" style="width: 100%; height:21px;" placeholder="Filter..." autocomplete="off">
    <div style="height: calc(100% - 23px); overflow: auto;">
      <div id="treetable" style="min-width: 100%;"></div>
    </div>
  </div>`;
    constructor() {
        super();
        let externalCss = document.createElement('style');
        externalCss.innerHTML = '@import url("./node_modules/jquery.fancytree/dist/skin-win8/ui.fancytree.css");';
        this.shadowRoot.appendChild(externalCss);
        this._filter = this._getDomElement('input');
        this._filter.onkeyup = () => {
            let match = this._filter.value;
            this._tree.filterNodes((node) => {
                return new RegExp(match, "i").test(node.title);
            });
        };
        this._treeDiv = this._getDomElement('treetable');
    }
    async ready() {
        $(this._treeDiv).fancytree({
            icon: true,
            extensions: ['childcounter', 'dnd5', 'filter'],
            quicksearch: true,
            source: [],
            dnd5: {
                dropMarkerParent: this.shadowRoot,
                preventRecursion: true,
                preventVoidMoves: false,
                dropMarkerOffsetX: -24,
                dropMarkerInsertOffsetX: -16,
                dragStart: (node, data) => {
                    data.effectAllowed = "all";
                    data.dataTransfer.setData(dragDropFormatNameElementDefinition, JSON.stringify(node.data.ref));
                    data.dropEffect = "copy";
                    return true;
                },
                dragEnter: (node, data) => {
                    return false;
                }
            }
        });
        //@ts-ignore
        this._tree = $.ui.fancytree.getTree(this._treeDiv);
        this._treeDiv.children[0].classList.add('fancytree-connectors');
    }
    async loadControls(serviceContainer, elementsServices) {
        let rootNode = this._tree.getRootNode();
        rootNode.removeChildren();
        for (const s of elementsServices) {
            const newNode = rootNode.addChildren({
                title: s.name,
                folder: true
            });
            let elements = await s.getElements();
            for (let e of elements) {
                newNode.addChildren({
                    title: e.name ?? e.tag,
                    folder: false,
                    //@ts-ignore
                    ref: e
                });
            }
            //@ts-ignore
            newNode.updateCounters();
        }
    }
}
customElements.define('node-projects-palette-tree-view', PaletteTreeView);
