// @node-projects/web-component-designer/dist/elements/services/selectionService/SelectionService.js
import { TypedEvent } from '/webui/node_modules/@node-projects/base-custom-webcomponent/./dist/index.js';
import { SelectionChangedAction } from '../undoService/transactionItems/SelectionChangedAction.js';
export class SelectionService {
    primarySelection;
    selectedElements = [];
    _designerCanvas;
    constructor(designerCanvas) {
        this._designerCanvas = designerCanvas;
    }
    setSelectedElements(designItems) {
        if (this.selectedElements != designItems) {
            const action = new SelectionChangedAction(this.selectedElements, designItems, this);
            this._designerCanvas.instanceServiceContainer.undoService.execute(action);
        }
    }
    _withoutUndoSetSelectedElements(designItems) {
        let oldSelectedElements = this.selectedElements;
        if (!designItems) {
            this.selectedElements = [];
            this.primarySelection = null;
        }
        else {
            let newSelection = [];
            for (let d of designItems) {
                if (d && d != d.instanceServiceContainer.contentService.rootDesignItem)
                    newSelection.push(d);
            }
            this.selectedElements = newSelection;
            if (newSelection && newSelection.length > 0)
                this.primarySelection = newSelection[0];
            else
                this.primarySelection = null;
        }
        this.onSelectionChanged.emit({ selectedElements: this.selectedElements, oldSelectedElements: oldSelectedElements });
    }
    clearSelectedElements() {
        this.setSelectedElements([]);
    }
    isSelected(designItem) {
        return this.selectedElements.indexOf(designItem) >= 0;
    }
    onSelectionChanged = new TypedEvent();
}
