// @node-projects/web-component-designer-visualization-addons/dist/components/EventAssignment.js
import { BaseCustomWebComponentConstructorAppend, css, html } from "@node-projects/base-custom-webcomponent";
import { ContextMenu, PropertiesHelper } from "@node-projects/web-component-designer";
import { ParameterEditor } from "./ParameterEditor.js";
import { BlocklyScriptEditor } from "../blockly/BlocklyScriptEditor.js";
import { SimpleScriptEditor } from "./SimpleScriptEditor.js";
export class EventAssignment extends BaseCustomWebComponentConstructorAppend {
    static style = css `
        :host {
            display: grid;
            grid-template-columns: 20px 1fr auto;
            overflow-y: auto;
            align-content: start;
            height: 100%;
        }
        .rect {
            width: 7px;
            height: 7px;
            border: 1px solid black;
            justify-self: center;
            cursor: pointer;
            align-self: center;
            white-space: nowrap;
        }
        input.mth {
            width: 100%;
            box-sizing: border-box;
        }
        input::placeholder {
            font-size: 8px;
        }
        a {
            cursor: pointer;
        }
        a:hover {
            cursor: pointer;
            text-decoration: underline;
        }
        button {
            cursor: pointer;
        }`;
    static template = html `
        <template repeat:item="[[this.events]]">
            <div @click="[[this._ctxMenu(event, item)]]" @contextmenu="[[this._ctxMenu(event, item)]]" class="rect" css:background-color="[[this._getScriptTypeColor(item)]]"></div>
            <a @click="[[this._showContextMenuAssignScript(event, item, false)]]" @contextmenu="[[this._ctxMenu(event, item)]]" title="[[item.name]]">[[item.name]]</a>
            <div>[[this._createControlsForScript(item)]]</div>
        </template>
        <span style="grid-column: 1 / span 3; margin-top: 8px; margin-left: 3px;">add event:</span>
        <input id="addEventInput" style="grid-column: 1 / span 3; margin: 5px;" @keypress="[[this._addEvent(event)]]" type="text">`;
    static editRowTemplate = html `
        <div style="display: flex; justify-content: flex-end;">
            <input value="[[this._getScriptName(item)]]" @keypress="[[this._changeScriptName(event, item)]]" hidden="[[this._getScriptType(item) !== 'js']]" placeholder="name" title="name" style="min-width: 50px; flex-basis: 30px; flex-grow: 2;" class="mth" type="text">
            <input value="[[this._getRelativeSignalsPath(item)]]" @keypress="[[this._changeRelativeSignalsPath(event, item)]]" placeholder="relative signals path" title="relative signals path" style="min-width: 50px; flex-basis: 20px; flex-grow: 1;" class="mth" type="text">
            <button css:background="[[this._hasParameters(item) ? 'lime' : '']]" style="display: flex; padding: 0; flex-grow: 0;" title="parameter" @click="[[this._editParameter(event, item)]]">p</button>
        </div>`;
    static scriptTypeColors = {
        'js': 'purple',
        'blockly': 'yellow',
        'script': 'lightgreen'
    };
    static is = 'node-projects-visualization-event-assignment';
    constructor() {
        super();
        this._restoreCachedInititalValues();
    }
    ready() {
        this._bindingsParse();
    }
    _instanceServiceContainer;
    _selectionChangedHandler;
    _selectedItems;
    _visualizationHandler;
    _visualizationShell;
    _scriptCommandsTypeInfo;
    _propertiesTypeInfo;
    events;
    initialize(visualizationHandler, visualizationShell, scriptCommandsTypeInfo, propertiesTypeInfo) {
        this._visualizationHandler = visualizationHandler;
        this._visualizationShell = visualizationShell;
        this._scriptCommandsTypeInfo = scriptCommandsTypeInfo;
        this._propertiesTypeInfo = propertiesTypeInfo;
    }
    set instanceServiceContainer(value) {
        this._instanceServiceContainer = value;
        this._selectionChangedHandler?.dispose();
        this._selectionChangedHandler = this._instanceServiceContainer.selectionService.onSelectionChanged.on(e => {
            this.selectedItems = e.selectedElements;
        });
        this.selectedItems = this._instanceServiceContainer.selectionService.selectedElements;
    }
    _createControlsForScript(eventItem) {
        const st = this._getScriptType(eventItem);
        switch (st) {
            case 'none':
                return '';
        }
        //@ts-ignore
        const ctl = this.constructor.editRowTemplate.content.cloneNode(true);
        return ctl;
    }
    _getScriptName(eventItem) {
        if (this.selectedItems[0].hasAttribute('@' + eventItem.name)) {
            const val = this.selectedItems[0].getAttribute('@' + eventItem.name);
            if (val[0] === '{') {
                if (val.includes('name')) {
                    const parsed = JSON.parse(val);
                    return parsed.name;
                }
            }
            else {
                return val;
            }
        }
        return null;
    }
    async _changeScriptName(e, eventItem) {
        if (e.key == 'Enter') {
            if (this.selectedItems && this.selectedItems.length) {
                if (this.selectedItems[0].hasAttribute('@' + eventItem.name)) {
                    const newValue = e.target.value;
                    const val = this.selectedItems[0].getAttribute('@' + eventItem.name);
                    if (val.startsWith('{')) {
                        const parsed = JSON.parse(val);
                        parsed.name = newValue;
                        this._selectedItems[0].setAttribute('@' + eventItem.name, JSON.stringify(parsed));
                    }
                    else {
                        this._selectedItems[0].setAttribute('@' + eventItem.name, newValue);
                    }
                }
            }
        }
    }
    _getRelativeSignalsPath(eventItem) {
        if (this.selectedItems[0].hasAttribute('@' + eventItem.name)) {
            const val = this.selectedItems[0].getAttribute('@' + eventItem.name);
            if (val[0] === '{' && val.includes('relativeSignalsPath')) {
                const parsed = JSON.parse(val);
                return parsed.relativeSignalsPath;
            }
        }
        return null;
    }
    async _changeRelativeSignalsPath(e, eventItem) {
        if (e.key == 'Enter') {
            if (this.selectedItems && this.selectedItems.length) {
                if (this.selectedItems[0].hasAttribute('@' + eventItem.name)) {
                    const newValue = e.target.value;
                    const val = this.selectedItems[0].getAttribute('@' + eventItem.name);
                    if (val.startsWith('{')) {
                        const parsed = JSON.parse(val);
                        parsed.relativeSignalsPath = newValue;
                        this._selectedItems[0].setAttribute('@' + eventItem.name, JSON.stringify(parsed));
                    }
                    else {
                        let obj = { name: val, relativeSignalsPath: newValue };
                        this._selectedItems[0].setAttribute('@' + eventItem.name, JSON.stringify(obj));
                    }
                }
            }
        }
    }
    _hasParameters(eventItem) {
        if (this.selectedItems[0].hasAttribute('@' + eventItem.name)) {
            const val = this.selectedItems[0].getAttribute('@' + eventItem.name);
            return val.includes('parameters');
        }
        return false;
    }
    _getScriptTypeColor(eventItem) {
        const st = this._getScriptType(eventItem);
        const color = EventAssignment.scriptTypeColors[st];
        return color ?? 'white';
    }
    _getScriptType(eventItem) {
        if (this.selectedItems && this.selectedItems.length) {
            if (this.selectedItems[0].hasAttribute('@' + eventItem.name)) {
                const val = this.selectedItems[0].getAttribute('@' + eventItem.name);
                if (val.startsWith('{')) {
                    const parsed = JSON.parse(val);
                    if ('blocks' in parsed)
                        return 'blockly';
                    if ('commands' in parsed)
                        return 'script';
                    return 'js';
                }
                else
                    return 'js';
            }
        }
        return 'none';
    }
    _getEventMethodname(eventItem) {
        if (this.selectedItems.length)
            return this.selectedItems[0].getAttribute('@' + eventItem.name);
        return '';
    }
    _inputMthName(event, eventItem) {
        let el = event.target;
        this.selectedItems[0].setAttribute('@' + eventItem.name, el.value);
    }
    _ctxMenu(e, eventItem) {
        e.preventDefault();
        const evtType = this._getScriptType(eventItem);
        if (evtType != 'none')
            ContextMenu.show([{ title: 'remove', action: () => { this.selectedItems[0].removeAttribute('@' + eventItem.name); this._bindingsRefresh(); } }], e);
        else
            this._showContextMenuAssignScript(e, eventItem, true);
    }
    async _addEvent(e) {
        if (e.key == 'Enter') {
            let ip = this._getDomElement('addEventInput');
            this._selectedItems[0].setAttribute('@' + PropertiesHelper.camelToDashCase(ip.value.replaceAll(' ', '-')), '');
            ip.value = '';
            this.scrollTop = 0;
            this.refresh();
        }
    }
    _createAssignScriptContextMenu(event, eventItem) {
        let ctxMenu = [
            {
                title: 'Simple Script',
                action: () => {
                    this._editEvent('script', event, eventItem);
                }
            },
            {
                title: 'Javascript',
                action: () => {
                    let nm = prompt('name of function ?');
                    if (nm) {
                        this._selectedItems[0].setAttribute('@' + eventItem.name, nm);
                        this.refresh();
                        this._editEvent('js', null, eventItem);
                    }
                }
            },
            {
                title: 'Blockly',
                action: () => {
                    this._editBlockly(null, eventItem);
                }
            }
        ];
        const evtType = this._getScriptType(eventItem);
        if (evtType != 'none') {
            ctxMenu.push({
                title: '-'
            });
            ctxMenu.push({
                title: 'remove',
                action: () => { this.selectedItems[0].removeAttribute('@' + eventItem.name); this._bindingsRefresh(); }
            });
        }
        return ctxMenu;
    }
    async _showContextMenuAssignScript(event, eventItem, isCtxMenu) {
        event.preventDefault();
        const evtType = this._getScriptType(eventItem);
        if (evtType != 'none' && !isCtxMenu) {
            this._editEvent(evtType, event, eventItem);
        }
        else {
            const ctxMenu = this._createAssignScriptContextMenu(event, eventItem);
            ContextMenu.show(ctxMenu, event);
        }
    }
    async _editParameter(e, eventItem) {
        const edt = new ParameterEditor();
        let existingParameter = {};
        edt.title = "ParameterEditor for '" + eventItem.name + "' of '" + this._selectedItems[0].name + "'";
        let data = this._selectedItems[0].getAttribute('@' + eventItem.name);
        if (data && data[0] == '{') {
            try {
                const parsed = JSON.parse(data);
                existingParameter = parsed.parameters;
            }
            catch { }
        }
        edt.setParametersObject(existingParameter);
        const result = await this._visualizationShell.openConfirmation(edt, { x: 100, y: 100, width: 700, height: 500 });
        if (result) {
            const par = edt.getParametersObject();
            let newObj = {
                name: data,
                parameters: par
            };
            if (data && data[0] == '{') {
                newObj = JSON.parse(data);
                newObj.parameters = par;
            }
            if (par == null)
                delete newObj.parameters;
            const newData = JSON.stringify(newObj);
            this._selectedItems[0].setAttribute('@' + eventItem.name, newData);
            this._bindingsRefresh();
        }
    }
    async _editBlockly(e, eventItem) {
        const edt = new BlocklyScriptEditor();
        edt.title = "Blockly Script for '" + eventItem.name + "' of '" + this._selectedItems[0].name + "'";
        let data = this._selectedItems[0].getAttribute('@' + eventItem.name);
        let parameters = null;
        let relativeSignalsPath = null;
        if (data) {
            const parsed = JSON.parse(data);
            parameters = parsed.parameters;
            relativeSignalsPath = parsed.relativeSignalsPath;
            edt.load(parsed);
        }
        const result = await this._visualizationShell.openConfirmation(edt, { x: 100, y: 100, width: 700, height: 500 });
        if (result) {
            const blockObj = edt.save();
            if (parameters) {
                blockObj.parameters = parameters;
            }
            if (relativeSignalsPath) {
                blockObj.relativeSignalsPath = relativeSignalsPath;
            }
            this._selectedItems[0].setAttribute('@' + eventItem.name, JSON.stringify(blockObj));
            this._bindingsRefresh();
        }
    }
    async _editJavascript(e, eventItem) {
        // todo ?
    }
    async _editSimpleScript(e, eventItem) {
        let scriptString = this.selectedItems[0].getAttribute('@' + eventItem.name);
        if (!scriptString || scriptString.startsWith('{')) {
            let script = { commands: [] };
            let parameters = null;
            let relativeSignalsPath = null;
            if (scriptString) {
                script = JSON.parse(scriptString);
                //@ts-ignore
                parameters = script.parameters;
                //@ts-ignore
                relativeSignalsPath = script.relativeSignalsPath;
            }
            let sc = new SimpleScriptEditor();
            sc.serviceContainer = this.selectedItems[0].serviceContainer;
            sc.scriptCommandsTypeInfo = this._scriptCommandsTypeInfo;
            sc.propertiesTypeInfo = this._propertiesTypeInfo;
            sc.visualizationShell = this._visualizationShell;
            sc.visualizationHandler = this._visualizationHandler;
            sc.loadScript(script);
            sc.title = "Script '" + eventItem.name + "' on " + this.selectedItems[0].name;
            let res = await this._visualizationShell.openConfirmation(sc, { x: 100, y: 100, width: 600, height: 500 });
            if (res) {
                let scriptCommands = sc.getScriptCommands();
                if (scriptCommands && scriptCommands.length) {
                    const sc = { commands: scriptCommands };
                    if (parameters) {
                        //@ts-ignore
                        sc.parameters = parameters;
                    }
                    if (relativeSignalsPath) {
                        //@ts-ignore
                        sc.relativeSignalsPath = relativeSignalsPath;
                    }
                    let json = JSON.stringify(sc);
                    this.selectedItems[0].setAttribute('@' + eventItem.name, json);
                    this._bindingsRefresh();
                }
            }
        }
    }
    async _editEvent(evtType, e, eventItem) {
        if (evtType == 'js') {
            this._editJavascript(e, eventItem);
        }
        else if (evtType == 'blockly') {
            this._editBlockly(e, eventItem);
        }
        else {
            this._editSimpleScript(e, eventItem);
        }
    }
    refresh() {
        if (this._selectedItems != null && this._selectedItems.length) {
            this.events = this._selectedItems[0].serviceContainer.getLastServiceWhere('eventsService', x => x.isHandledElement(this._selectedItems[0])).getPossibleEvents(this._selectedItems[0]);
        }
        else {
            this.events = [];
        }
        this._bindingsRefresh();
    }
    get selectedItems() {
        return this._selectedItems;
    }
    set selectedItems(items) {
        if (this._selectedItems != items) {
            this._selectedItems = items;
            this.refresh();
        }
    }
}
customElements.define(EventAssignment.is, EventAssignment);
