// @node-projects/web-component-designer/dist/elements/services/collaborationService/WebRtcTabCollaborationTransport.js
export const defaultWebRtcTabCollaborationSignalingChannels = ['broadcast-channel', 'manual'];
function createRandomId() {
    return globalThis.crypto?.randomUUID?.() ?? `wcd-collab-${Date.now()}-${Math.floor(Math.random() * 100000)}`;
}
function createEnabledSignalingChannels(channels) {
    if (channels == null)
        return new Set(defaultWebRtcTabCollaborationSignalingChannels);
    const validChannels = channels.filter((channel) => channel === 'broadcast-channel' || channel === 'manual');
    return new Set(validChannels);
}
function areSetsEqual(first, second) {
    if (first.size !== second.size)
        return false;
    for (const entry of first) {
        if (!second.has(entry))
            return false;
    }
    return true;
}
function cloneSignalingMessage(message) {
    return {
        ...message,
        description: message.description ? { ...message.description } : undefined,
        candidate: message.candidate ? { ...message.candidate } : undefined
    };
}
function escapeControlCharacter(character) {
    switch (character) {
        case '\b':
            return '\\b';
        case '\f':
            return '\\f';
        case '\n':
            return '\\n';
        case '\r':
            return '\\r';
        case '\t':
            return '\\t';
        default:
            return `\\u${character.charCodeAt(0).toString(16).padStart(4, '0')}`;
    }
}
function normalizeJsonControlCharactersInStrings(data) {
    let normalized = '';
    let inString = false;
    let isEscaped = false;
    for (const character of data) {
        if (inString) {
            if (isEscaped) {
                normalized += character;
                isEscaped = false;
                continue;
            }
            if (character === '\\') {
                normalized += character;
                isEscaped = true;
                continue;
            }
            if (character === '"') {
                normalized += character;
                inString = false;
                continue;
            }
            if (character < ' ') {
                normalized += escapeControlCharacter(character);
                continue;
            }
            normalized += character;
            continue;
        }
        normalized += character;
        if (character === '"')
            inString = true;
    }
    return normalized;
}
function parseManualSignalingBundleData(data) {
    try {
        return JSON.parse(data);
    }
    catch (error) {
        const normalizedData = normalizeJsonControlCharactersInStrings(data);
        if (normalizedData === data)
            throw error;
        return JSON.parse(normalizedData);
    }
}
export class WebRtcTabCollaborationTransport {
    _service = null;
    _session = null;
    _broadcastSignalingChannel = null;
    _enabledSignalingChannels = createEnabledSignalingChannels();
    _manualSignalingMessages = [];
    _peerConnections = new Map();
    _dataChannels = new Map();
    _initiatedPeers = new Set();
    _peersNeedingInitialSnapshot = new Set();
    _pendingCandidates = new Map();
    _processedSignalIds = new Set();
    _remoteConnectionIds = new Map();
    _localConnectionId = null;
    constructor(options) {
        this._enabledSignalingChannels = createEnabledSignalingChannels(options?.enabledSignalingChannels);
    }
    get enabledSignalingChannels() {
        return [...this._enabledSignalingChannels];
    }
    setEnabledSignalingChannels(channels) {
        const nextChannels = createEnabledSignalingChannels(channels);
        if (areSetsEqual(this._enabledSignalingChannels, nextChannels))
            return;
        this._enabledSignalingChannels = nextChannels;
        this.ensureBroadcastChannelState();
        if (this._session && this.hasAvailableSignalingChannel()) {
            this.sendSignal({
                sessionId: this._session.sessionId,
                fromPeerId: this._session.peerId,
                type: 'hello',
                requestInitialSnapshot: this.shouldRequestInitialSnapshot()
            });
        }
    }
    getManualSignalingBundle() {
        if (!this._session || !this._localConnectionId)
            return null;
        return {
            version: 1,
            sessionId: this._session.sessionId,
            fromPeerId: this._session.peerId,
            connectionId: this._localConnectionId,
            createdAt: Date.now(),
            messages: this._manualSignalingMessages.map(cloneSignalingMessage)
        };
    }
    exportManualSignalingData(space = 2) {
        const bundle = this.getManualSignalingBundle();
        return bundle ? JSON.stringify(bundle, null, space) : '';
    }
    async importManualSignalingData(data) {
        if (!this._session)
            throw new Error('Collaboration is not connected for this document.');
        const bundle = typeof data === 'string'
            ? parseManualSignalingBundleData(data)
            : data;
        if (!bundle || !Array.isArray(bundle.messages))
            throw new Error('Invalid collaboration signaling data.');
        if (bundle.sessionId !== this._session.sessionId)
            throw new Error(`Signaling data is for session "${bundle.sessionId}", but the active session is "${this._session.sessionId}".`);
        let importedCount = 0;
        let ignoredCount = 0;
        for (const message of bundle.messages) {
            if (!this.isSignalingMessage(message)) {
                ignoredCount++;
                continue;
            }
            const handled = await this.handleSignalMessage(message);
            if (handled)
                importedCount++;
            else
                ignoredCount++;
        }
        return { importedCount, ignoredCount };
    }
    async attach(service) {
        this._service = service;
    }
    async detach() {
        await this.disconnect();
        this._service = null;
    }
    async connect(session) {
        if (typeof RTCPeerConnection === 'undefined') {
            console.warn('WebRTC collaboration is not available in this browser.');
            return;
        }
        await this.disconnect();
        this._session = session;
        this._localConnectionId = createRandomId();
        this.ensureBroadcastChannelState();
        if (!this.hasAvailableSignalingChannel())
            console.warn('WebRTC collaboration signaling is disabled. Enable manual or broadcast signaling to connect peers.');
        this.sendSignal({
            sessionId: session.sessionId,
            fromPeerId: session.peerId,
            type: 'hello',
            requestInitialSnapshot: this.shouldRequestInitialSnapshot()
        });
    }
    async disconnect() {
        if (this._session) {
            this.sendSignal({
                sessionId: this._session.sessionId,
                fromPeerId: this._session.peerId,
                type: 'bye'
            }, { includeManual: false });
        }
        if (this._broadcastSignalingChannel) {
            this._broadcastSignalingChannel.close();
            this._broadcastSignalingChannel = null;
        }
        for (const channel of this._dataChannels.values()) {
            try {
                channel.close();
            }
            catch {
            }
        }
        this._dataChannels.clear();
        for (const peerConnection of this._peerConnections.values()) {
            try {
                peerConnection.close();
            }
            catch {
            }
        }
        this._peerConnections.clear();
        this._pendingCandidates.clear();
        this._initiatedPeers.clear();
        this._peersNeedingInitialSnapshot.clear();
        this._manualSignalingMessages = [];
        this._processedSignalIds.clear();
        this._remoteConnectionIds.clear();
        this._localConnectionId = null;
        this._session = null;
    }
    async sendChange(change, snapshot) {
        this.sendData({ type: 'change', change, snapshot });
    }
    async sendSelection(selection) {
        this.sendData({ type: 'selection', selection });
    }
    async sendPresence(peer) {
        this.sendData({ type: 'presence', peer });
    }
    async sendComment(change) {
        if (change.comment)
            this.sendData({ type: 'comment-upsert', comment: change.comment });
        else if (change.commentId)
            this.sendData({ type: 'comment-remove', commentId: change.commentId });
    }
    ensureBroadcastChannelState() {
        if (!this._session || !this._enabledSignalingChannels.has('broadcast-channel') || typeof BroadcastChannel === 'undefined') {
            if (this._broadcastSignalingChannel) {
                this._broadcastSignalingChannel.close();
                this._broadcastSignalingChannel = null;
            }
            return;
        }
        if (this._broadcastSignalingChannel)
            return;
        this._broadcastSignalingChannel = new BroadcastChannel(`wcd-collab:${this._session.sessionId}`);
        this._broadcastSignalingChannel.onmessage = event => {
            void this.handleSignalMessage(event.data);
        };
    }
    hasAvailableSignalingChannel() {
        return this._enabledSignalingChannels.has('manual')
            || (this._enabledSignalingChannels.has('broadcast-channel') && typeof BroadcastChannel !== 'undefined');
    }
    isSignalingMessage(message) {
        if (!message)
            return false;
        return typeof message.id === 'string'
            && typeof message.connectionId === 'string'
            && typeof message.createdAt === 'number'
            && typeof message.sessionId === 'string'
            && typeof message.fromPeerId === 'string'
            && (message.requestInitialSnapshot == null || typeof message.requestInitialSnapshot === 'boolean')
            && (message.type === 'hello'
                || message.type === 'hello-ack'
                || message.type === 'offer'
                || message.type === 'answer'
                || message.type === 'ice'
                || message.type === 'bye');
    }
    rememberRemoteConnection(remotePeerId, connectionId) {
        const previousConnectionId = this._remoteConnectionIds.get(remotePeerId);
        if (previousConnectionId === connectionId)
            return;
        this.cleanupPeer(remotePeerId, false);
        this._remoteConnectionIds.set(remotePeerId, connectionId);
    }
    shouldRequestInitialSnapshot() {
        const snapshot = this._service?.createSnapshot();
        return !snapshot?.html?.trim();
    }
    updateInitialSnapshotRequest(remotePeerId, requestInitialSnapshot) {
        if (requestInitialSnapshot)
            this._peersNeedingInitialSnapshot.add(remotePeerId);
        else
            this._peersNeedingInitialSnapshot.delete(remotePeerId);
    }
    async handleSignalMessage(message) {
        if (!this._session || !message || message.sessionId !== this._session.sessionId || message.fromPeerId === this._session.peerId)
            return false;
        if (message.toPeerId && message.toPeerId !== this._session.peerId)
            return false;
        if (this._processedSignalIds.has(message.id))
            return false;
        this._processedSignalIds.add(message.id);
        const knownRemoteConnectionId = this._remoteConnectionIds.get(message.fromPeerId);
        const isHandshakeMessage = message.type === 'hello' || message.type === 'hello-ack';
        if (!isHandshakeMessage) {
            if (!knownRemoteConnectionId || knownRemoteConnectionId !== message.connectionId)
                return false;
        }
        switch (message.type) {
            case 'hello':
                this.rememberRemoteConnection(message.fromPeerId, message.connectionId);
                this.updateInitialSnapshotRequest(message.fromPeerId, message.requestInitialSnapshot);
                this.sendSignal({
                    sessionId: this._session.sessionId,
                    fromPeerId: this._session.peerId,
                    toPeerId: message.fromPeerId,
                    type: 'hello-ack',
                    requestInitialSnapshot: this.shouldRequestInitialSnapshot()
                });
                if (this.shouldInitiate(message.fromPeerId))
                    await this.ensureOffer(message.fromPeerId);
                else
                    this.getOrCreatePeerConnection(message.fromPeerId);
                return true;
            case 'hello-ack':
                this.rememberRemoteConnection(message.fromPeerId, message.connectionId);
                this.updateInitialSnapshotRequest(message.fromPeerId, message.requestInitialSnapshot);
                if (this.shouldInitiate(message.fromPeerId))
                    await this.ensureOffer(message.fromPeerId);
                return true;
            case 'offer':
                if (message.description)
                    await this.acceptOffer(message.fromPeerId, message.description);
                return true;
            case 'answer':
                if (message.description)
                    await this.acceptAnswer(message.fromPeerId, message.description);
                return true;
            case 'ice':
                if (message.candidate)
                    await this.acceptIceCandidate(message.fromPeerId, message.candidate);
                return true;
            case 'bye':
                this.cleanupPeer(message.fromPeerId, false);
                this._remoteConnectionIds.delete(message.fromPeerId);
                this._service?.removePeer(message.fromPeerId, 'remote');
                return true;
        }
    }
    shouldInitiate(remotePeerId) {
        if (!this._session)
            return false;
        return this._session.peerId.localeCompare(remotePeerId) < 0;
    }
    getOrCreatePeerConnection(remotePeerId) {
        let peerConnection = this._peerConnections.get(remotePeerId);
        if (peerConnection)
            return peerConnection;
        peerConnection = new RTCPeerConnection();
        peerConnection.onicecandidate = event => {
            if (event.candidate && this._session) {
                this.sendSignal({
                    sessionId: this._session.sessionId,
                    fromPeerId: this._session.peerId,
                    toPeerId: remotePeerId,
                    type: 'ice',
                    candidate: event.candidate.toJSON()
                });
            }
        };
        peerConnection.ondatachannel = event => {
            this.registerDataChannel(remotePeerId, event.channel);
        };
        peerConnection.onconnectionstatechange = () => {
            if (peerConnection.connectionState === 'closed' || peerConnection.connectionState === 'failed' || peerConnection.connectionState === 'disconnected') {
                this.cleanupPeer(remotePeerId, false);
                this._service?.removePeer(remotePeerId, 'remote');
            }
        };
        this._peerConnections.set(remotePeerId, peerConnection);
        return peerConnection;
    }
    async ensureOffer(remotePeerId) {
        if (this._initiatedPeers.has(remotePeerId))
            return;
        const peerConnection = this.getOrCreatePeerConnection(remotePeerId);
        this._initiatedPeers.add(remotePeerId);
        if (!this._dataChannels.has(remotePeerId)) {
            const channel = peerConnection.createDataChannel('web-component-designer-collaboration');
            this.registerDataChannel(remotePeerId, channel);
        }
        const offer = await peerConnection.createOffer();
        await peerConnection.setLocalDescription(offer);
        if (!this._session)
            return;
        this.sendSignal({
            sessionId: this._session.sessionId,
            fromPeerId: this._session.peerId,
            toPeerId: remotePeerId,
            type: 'offer',
            description: peerConnection.localDescription?.toJSON() ?? offer
        });
    }
    async acceptOffer(remotePeerId, description) {
        const peerConnection = this.getOrCreatePeerConnection(remotePeerId);
        await peerConnection.setRemoteDescription(description);
        await this.flushPendingCandidates(remotePeerId);
        const answer = await peerConnection.createAnswer();
        await peerConnection.setLocalDescription(answer);
        if (!this._session)
            return;
        this.sendSignal({
            sessionId: this._session.sessionId,
            fromPeerId: this._session.peerId,
            toPeerId: remotePeerId,
            type: 'answer',
            description: peerConnection.localDescription?.toJSON() ?? answer
        });
    }
    async acceptAnswer(remotePeerId, description) {
        const peerConnection = this.getOrCreatePeerConnection(remotePeerId);
        await peerConnection.setRemoteDescription(description);
        await this.flushPendingCandidates(remotePeerId);
    }
    async acceptIceCandidate(remotePeerId, candidate) {
        const peerConnection = this.getOrCreatePeerConnection(remotePeerId);
        if (!peerConnection.remoteDescription) {
            const pending = this._pendingCandidates.get(remotePeerId) ?? [];
            pending.push(candidate);
            this._pendingCandidates.set(remotePeerId, pending);
            return;
        }
        await peerConnection.addIceCandidate(candidate);
    }
    async flushPendingCandidates(remotePeerId) {
        const peerConnection = this._peerConnections.get(remotePeerId);
        const candidates = this._pendingCandidates.get(remotePeerId);
        if (!peerConnection || !candidates?.length)
            return;
        for (const candidate of candidates)
            await peerConnection.addIceCandidate(candidate);
        this._pendingCandidates.delete(remotePeerId);
    }
    registerDataChannel(remotePeerId, channel) {
        const previousChannel = this._dataChannels.get(remotePeerId);
        if (previousChannel && previousChannel !== channel) {
            try {
                previousChannel.close();
            }
            catch {
            }
        }
        channel.onopen = () => {
            void this.handleOpenChannel(remotePeerId);
        };
        channel.onmessage = event => {
            void this.handleDataMessage(remotePeerId, event.data);
        };
        channel.onclose = () => {
            if (this._dataChannels.get(remotePeerId) === channel)
                this._dataChannels.delete(remotePeerId);
        };
        this._dataChannels.set(remotePeerId, channel);
    }
    async handleOpenChannel(remotePeerId) {
        if (!this._session || !this._service)
            return;
        const session = this._session;
        const service = this._service;
        this._manualSignalingMessages = this._manualSignalingMessages.filter(x => x.toPeerId !== remotePeerId);
        const localPeer = service.peers.find(x => x.peerId === session.peerId);
        if (localPeer)
            await this.sendPresence(localPeer);
        const currentSelection = localPeer?.selectedNodeIndexes?.length || localPeer?.activeNodeIndex != null
            ? {
                peerId: session.peerId,
                selectedNodeIndexes: [...(localPeer.selectedNodeIndexes ?? [])],
                primaryNodeIndex: localPeer.activeNodeIndex
            }
            : null;
        if (currentSelection)
            await this.sendSelection(currentSelection);
        if (this._peersNeedingInitialSnapshot.has(remotePeerId)) {
            this._peersNeedingInitialSnapshot.delete(remotePeerId);
            this.sendData({ type: 'snapshot', snapshot: service.createSnapshot() }, remotePeerId);
            for (const comment of service.comments)
                this.sendData({ type: 'comment-upsert', comment }, remotePeerId);
        }
    }
    async handleDataMessage(remotePeerId, rawData) {
        if (!this._service)
            return;
        const message = JSON.parse(rawData);
        switch (message.type) {
            case 'snapshot':
                await this._service.applyRemoteSnapshot(message.snapshot);
                break;
            case 'change':
                await this._service.applyRemoteChange(message.change, message.snapshot);
                break;
            case 'selection':
                this._service.updateRemoteSelection(remotePeerId, message.selection.selectedNodeIndexes ?? [], message.selection.primaryNodeIndex);
                break;
            case 'presence':
                this._service.updatePeerPresence({ ...message.peer, peerId: remotePeerId }, 'remote');
                break;
            case 'comment-upsert':
                this._service.upsertComment(message.comment, 'remote');
                break;
            case 'comment-remove':
                this._service.removeComment(message.commentId, 'remote');
                break;
        }
    }
    sendSignal(message, options) {
        if (!this._localConnectionId)
            return;
        const signal = {
            ...message,
            id: createRandomId(),
            createdAt: Date.now(),
            connectionId: this._localConnectionId
        };
        this._processedSignalIds.add(signal.id);
        const includeBroadcast = options?.includeBroadcast ?? this._enabledSignalingChannels.has('broadcast-channel');
        if (includeBroadcast)
            this._broadcastSignalingChannel?.postMessage(signal);
        const includeManual = options?.includeManual ?? this._enabledSignalingChannels.has('manual');
        if (includeManual && signal.type !== 'bye')
            this.enqueueManualSignal(signal);
    }
    enqueueManualSignal(signal) {
        if (signal.type === 'hello')
            this._manualSignalingMessages = this._manualSignalingMessages.filter(x => x.type !== 'hello');
        else if (signal.type === 'hello-ack' || signal.type === 'offer' || signal.type === 'answer')
            this._manualSignalingMessages = this._manualSignalingMessages.filter(x => !(x.type === signal.type && x.toPeerId === signal.toPeerId));
        this._manualSignalingMessages.push(cloneSignalingMessage(signal));
    }
    sendData(message, remotePeerId) {
        const payload = JSON.stringify(message);
        for (const [peerId, channel] of this._dataChannels) {
            if (remotePeerId && peerId !== remotePeerId)
                continue;
            if (channel.readyState === 'open')
                channel.send(payload);
        }
    }
    cleanupPeer(remotePeerId, clearInitialSnapshotState = true) {
        const dataChannel = this._dataChannels.get(remotePeerId);
        if (dataChannel) {
            try {
                dataChannel.close();
            }
            catch {
            }
            this._dataChannels.delete(remotePeerId);
        }
        const peerConnection = this._peerConnections.get(remotePeerId);
        if (peerConnection) {
            try {
                peerConnection.close();
            }
            catch {
            }
            this._peerConnections.delete(remotePeerId);
        }
        this._initiatedPeers.delete(remotePeerId);
        if (clearInitialSnapshotState)
            this._peersNeedingInitialSnapshot.delete(remotePeerId);
        this._pendingCandidates.delete(remotePeerId);
    }
}
